// @rschedule/core/umd/generators.js
(function(global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined'
    ? factory(exports, require('@rschedule/core'))
    : typeof define === 'function' && define.amd
    ? define(['exports', '@rschedule/core'], factory)
    : ((global = global || self), factory((global.rScheduleGenerators = {}), global.rSchedule));
})(this, function(exports, core) {
  'use strict';

  /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
  /* global Reflect, Promise */

  var extendStatics = function(d, b) {
    extendStatics =
      Object.setPrototypeOf ||
      ({ __proto__: [] } instanceof Array &&
        function(d, b) {
          d.__proto__ = b;
        }) ||
      function(d, b) {
        for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
      };
    return extendStatics(d, b);
  };

  function __extends(d, b) {
    extendStatics(d, b);
    function __() {
      this.constructor = d;
    }
    d.prototype = b === null ? Object.create(b) : ((__.prototype = b.prototype), new __());
  }

  var __assign = function() {
    __assign =
      Object.assign ||
      function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
          s = arguments[i];
          for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
      };
    return __assign.apply(this, arguments);
  };

  function __generator(thisArg, body) {
    var _ = {
        label: 0,
        sent: function() {
          if (t[0] & 1) throw t[1];
          return t[1];
        },
        trys: [],
        ops: [],
      },
      f,
      y,
      t,
      g;
    return (
      (g = { next: verb(0), throw: verb(1), return: verb(2) }),
      typeof Symbol === 'function' &&
        (g[Symbol.iterator] = function() {
          return this;
        }),
      g
    );
    function verb(n) {
      return function(v) {
        return step([n, v]);
      };
    }
    function step(op) {
      if (f) throw new TypeError('Generator is already executing.');
      while (_)
        try {
          if (
            ((f = 1),
            y &&
              (t =
                op[0] & 2
                  ? y['return']
                  : op[0]
                  ? y['throw'] || ((t = y['return']) && t.call(y), 0)
                  : y.next) &&
              !(t = t.call(y, op[1])).done)
          )
            return t;
          if (((y = 0), t)) op = [op[0] & 2, t.value];
          switch (op[0]) {
            case 0:
            case 1:
              t = op;
              break;
            case 4:
              _.label++;
              return { value: op[1], done: false };
            case 5:
              _.label++;
              y = op[1];
              op = [0];
              continue;
            case 7:
              op = _.ops.pop();
              _.trys.pop();
              continue;
            default:
              if (
                !((t = _.trys), (t = t.length > 0 && t[t.length - 1])) &&
                (op[0] === 6 || op[0] === 2)
              ) {
                _ = 0;
                continue;
              }
              if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                _.label = op[1];
                break;
              }
              if (op[0] === 6 && _.label < t[1]) {
                _.label = t[1];
                t = op;
                break;
              }
              if (t && _.label < t[2]) {
                _.label = t[2];
                _.ops.push(op);
                break;
              }
              if (t[2]) _.ops.pop();
              _.trys.pop();
              continue;
          }
          op = body.call(thisArg, _);
        } catch (e) {
          op = [6, e];
          y = 0;
        } finally {
          f = t = 0;
        }
      if (op[0] & 5) throw op[1];
      return { value: op[0] ? op[1] : void 0, done: true };
    }
  }

  function __values(o) {
    var m = typeof Symbol === 'function' && o[Symbol.iterator],
      i = 0;
    if (m) return m.call(o);
    return {
      next: function() {
        if (o && i >= o.length) o = void 0;
        return { value: o && o[i++], done: !o };
      },
    };
  }

  function __read(o, n) {
    var m = typeof Symbol === 'function' && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o),
      r,
      ar = [],
      e;
    try {
      while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    } catch (error) {
      e = { error: error };
    } finally {
      try {
        if (r && !r.done && (m = i['return'])) m.call(i);
      } finally {
        if (e) throw e.error;
      }
    }
    return ar;
  }

  function __spread() {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
  }

  var OccurrenceGenerator = /** @class */ (function() {
    function OccurrenceGenerator(args) {
      if (args === void 0) {
        args = {};
      }
      this.timezone = args.timezone !== undefined ? args.timezone : null;
      this.maxDuration = args.maxDuration || 0;
    }
    Object.defineProperty(OccurrenceGenerator.prototype, 'firstDate', {
      /** Returns the first occurrence or, if there are no occurrences, null. */
      get: function() {
        if (this._firstDate !== undefined) return this._firstDate;
        var start = this._run().next().value;
        this._firstDate = start ? this.dateAdapter.fromDateTime(start) : null;
        return this._firstDate;
      },
      enumerable: false,
      configurable: true,
    });
    Object.defineProperty(OccurrenceGenerator.prototype, 'lastDate', {
      /** If generator is infinite, returns `null`. Otherwise returns the end date */
      get: function() {
        if (this._lastDate !== undefined) return this._lastDate;
        if (this.isInfinite) {
          this._lastDate = null;
          return null;
        }
        var end = this._run({ reverse: true }).next().value;
        this._lastDate = end ? this.dateAdapter.fromDateTime(end) : null;
        return this._lastDate;
      },
      enumerable: false,
      configurable: true,
    });
    OccurrenceGenerator.prototype.pipe = function() {
      var _this = this;
      var operators = [];
      for (var _i = 0; _i < arguments.length; _i++) {
        operators[_i] = arguments[_i];
      }
      return operators.reduce(function(prev, curr) {
        return curr({ base: prev, timezone: _this.timezone });
      }, this);
    };
    /**
         * Processes the object's rules/dates and returns an iterable for the occurrences.
         *
         * Options object:
         * - `start` the date to begin iteration on
         * - `end` the date to end iteration on
         * - `take` the max number of dates to take before ending iteration
         * - `reverse` whether to iterate in reverse or not
         *
         * Examples:
         *
         * ```
         * const iterator = schedule.occurrences({ start: new Date(), take: 5 });
         
         * for (const date of iterator) {
         *   // do stuff
         * }
      
         * iterator.toArray() // returns Date array
         * iterator.next().value // returns next Date
         * ```
         *
         */
    OccurrenceGenerator.prototype.occurrences = function(args) {
      if (args === void 0) {
        args = {};
      }
      return new OccurrenceIterator(this, this.normalizeOccurrencesArgs(args));
    };
    /**
     * Iterates over the object's occurrences and bundles them into collections
     * with a specified granularity (default is `"YEARLY"`). Make sure to
     * read about each option & combination of options below.
     *
     * Options object:
     *   - start?: DateAdapter
     *   - end?: DateAdapter
     *   - take?: number
     *   - reverse?: NOT SUPPORTED
     *   - granularity?: CollectionsGranularity
     *   - weekStart?: DateAdapter.Weekday
     *   - skipEmptyPeriods?: boolean
     *
     * Returned `Collection` object:
     *
     *   - `dates` property containing an array of DateAdapter objects.
     *   - `granularity` property containing the granularity.
     *     - `CollectionsGranularity` === `RuleOptions.Frequency`.
     *     - default is `"YEARLY"`
     *   - `periodStart` property containing a DateAdapter equal to the period's
     *     start time.
     *   - `periodEnd` property containing a DateAdapter equal to the period's
     *     end time.
     *
     * #### Details:
     *
     * `collections()` always returns full periods. This means that the `start` argument is
     * transformed to be the start of whatever period the `start` argument is in, and the
     * `end` argument is transformed to be the end of whatever period the `end` argument is
     * in.
     *
     * - Example: with granularity `"YEARLY"`, the `start` argument will be transformed to be the
     *   start of the year passed in the `start` argument, and the `end` argument will be transformed
     *   to be the end of the year passed in the `end` argument.
     *
     * By default, the `periodStart` value of `Collection` objects produced by this method increments linearly.
     * This means the returned `Collection#dates` property may have length 0. This can be changed by
     * passing the `skipEmptyPeriods: true` option, in which case the `periodStart` from one collection to the
     * next can "jump".
     *
     * - Example 1: if your object's first occurrence is 2019/2/1 (February 1st) and you call
     *   `collection({skipEmptyPeriods: true, granularity: 'DAILY', start: new Date(2019,0,1)})`
     *   (so starting on January 1st), the first Collection produced will have a `periodStart` in February.
     *
     * - Example 2: if your object's first occurrence is 2019/2/1 (February 1st) and you call
     *   `collection({granularity: 'DAILY', start: new Date(2019,0,1)})`
     *   (so starting on January 1st), the first collection produced will have a `Collection#periodStart`
     *   of January 1st and have `Collection#dates === []`. Similarly, the next 30 collections produced
     *   (Jan 2nd - 31st) will all contain an empty array for the `dates` property. Then the February 1st
     *   `Collection` will contain dates.
     *
     * When giving a `take` argument to `collections()`, you are specifying
     * the number of `Collection` objects to return (rather than occurrences).
     *
     * When choosing a granularity of `"WEEKLY"`, the `weekStart` option is required.
     *
     * When choosing a granularity of `"MONTHLY"`:
     *
     * - If the `weekStart` option *is not* present, will generate collections with
     *   the `periodStart` and `periodEnd` at the beginning and end of each month.
     *
     * - If the `weekStart` option *is* present, will generate collections with the
     *   `periodStart` equal to the start of the first week of the month, and the
     *   `periodEnd` equal to the end of the last week of the month. This behavior could be
     *   desired when rendering opportunities in a calendar view, where the calendar renders
     *   full weeks (which may result in the calendar displaying dates in the
     *   previous or next months).
     *
     */
    OccurrenceGenerator.prototype.collections = function(args) {
      if (args === void 0) {
        args = {};
      }
      return new CollectionIterator(this, this.normalizeCollectionsArgs(args));
    };
    /**
     * Returns true if an occurrence starts on or between the provided start/end
     * datetimes. If the `excludeEnds` option is provided, then occurrences
     * equal to the start/end times are ignored.
     *
     * If the occurrence generator has a duration, and `excludeEnds !== true`,
     * and a `maxDuration` argument is supplied (either in the constructor or
     * here), then any occurrence that's time overlaps with the start/end times
     * return true.
     */
    OccurrenceGenerator.prototype.occursBetween = function(startInput, endInput, options) {
      var e_1, _a, e_2, _b;
      if (options === void 0) {
        options = {};
      }
      var start = this.normalizeDateInput(startInput);
      var end = this.normalizeDateInput(endInput);
      if (this.hasDuration && !options.excludeEnds) {
        var maxDuration = this.getMaxDuration('occursBetween', options);
        var iterator = this._run({
          start: start.subtract(maxDuration, 'millisecond'),
          end: end,
          reverse: true,
        });
        try {
          for (
            var iterator_1 = __values(iterator), iterator_1_1 = iterator_1.next();
            !iterator_1_1.done;
            iterator_1_1 = iterator_1.next()
          ) {
            var day = iterator_1_1.value;
            if (day.end.isBefore(start)) continue;
            return true;
          }
        } catch (e_1_1) {
          e_1 = { error: e_1_1 };
        } finally {
          try {
            if (iterator_1_1 && !iterator_1_1.done && (_a = iterator_1.return)) _a.call(iterator_1);
          } finally {
            if (e_1) throw e_1.error;
          }
        }
        return false;
      }
      try {
        for (
          var _c = __values(this._run({ start: start, end: end })), _d = _c.next();
          !_d.done;
          _d = _c.next()
        ) {
          var day = _d.value;
          if (options.excludeEnds) {
            if (day.isEqual(start)) {
              continue;
            }
            if (day.isEqual(end)) {
              break;
            }
          }
          return true;
        }
      } catch (e_2_1) {
        e_2 = { error: e_2_1 };
      } finally {
        try {
          if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
        } finally {
          if (e_2) throw e_2.error;
        }
      }
      return false;
    };
    OccurrenceGenerator.prototype.occursOn = function(rawArgs) {
      var e_3, _a, e_4, _b;
      var args = this.normalizeOccursOnArgs(rawArgs);
      if (args.weekday) {
        if (this.isInfinite && !args.before) {
          throw new core.ArgumentError(
            'When calling `occursOn()` with a `weekday` argument ' +
              'and an occurrence object that has infinite occurrences, ' +
              'you must include a `before` argument as well.',
          );
        }
        var start = args.after && (args.excludeEnds ? args.after.add(1, 'day') : args.after);
        var end = args.before && (args.excludeEnds ? args.before.subtract(1, 'day') : args.before);
        var iterator = this._run({ start: start, end: end });
        var date = iterator.next().value;
        if (!date) return false;
        while (date) {
          if (date.get('weekday') === args.weekday) {
            return true;
          }
          date = iterator.next({
            skipToDate: date
              .add(core.getDifferenceBetweenWeekdays(date.get('weekday'), args.weekday), 'day')
              .granularity('day'),
          }).value;
        }
        return false;
      }
      if (!args.date) {
        throw new core.ArgumentError(
          '"occursOn()" must be called with either a "date" or "weekday" argument',
        );
      }
      if (this.hasDuration) {
        var maxDuration = this.getMaxDuration('occursOn', args);
        var iterator = this._run({
          start: args.date.subtract(maxDuration, 'millisecond'),
          end: args.date,
        });
        try {
          for (
            var iterator_2 = __values(iterator), iterator_2_1 = iterator_2.next();
            !iterator_2_1.done;
            iterator_2_1 = iterator_2.next()
          ) {
            var date = iterator_2_1.value;
            if (date.end.isBefore(args.date)) continue;
            if (date.isAfter(args.date)) return false;
            return true;
          }
        } catch (e_3_1) {
          e_3 = { error: e_3_1 };
        } finally {
          try {
            if (iterator_2_1 && !iterator_2_1.done && (_a = iterator_2.return)) _a.call(iterator_2);
          } finally {
            if (e_3) throw e_3.error;
          }
        }
        return false;
      }
      try {
        for (
          var _c = __values(this._run({ start: args.date, end: args.date })), _d = _c.next();
          !_d.done;
          _d = _c.next()
        ) {
          var day = _d.value;
          return !!day;
        }
      } catch (e_4_1) {
        e_4 = { error: e_4_1 };
      } finally {
        try {
          if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
        } finally {
          if (e_4) throw e_4.error;
        }
      }
      return false;
    };
    /**
     * Returns true if an occurrence starts after the provided datetime.
     * If the `excludeStart` option is provided, then occurrences
     * equal to the provided datetime are ignored.
     *
     * If the occurrence generator has a duration, and `excludeStart !== true`,
     * and a `maxDuration` argument is supplied (either in the constructor or
     * here), then any occurrence that's end time is after/equal to the provided
     * datetime return true.
     */
    OccurrenceGenerator.prototype.occursAfter = function(date, options) {
      var e_5, _a, e_6, _b;
      if (options === void 0) {
        options = {};
      }
      var adapter = this.normalizeDateInput(date);
      if (this.hasDuration && !options.excludeStart) {
        var maxDuration = this.getMaxDuration('occursAfter', options);
        var iterator = this._run({
          start: adapter.subtract(maxDuration, 'millisecond'),
        });
        try {
          for (
            var iterator_3 = __values(iterator), iterator_3_1 = iterator_3.next();
            !iterator_3_1.done;
            iterator_3_1 = iterator_3.next()
          ) {
            var date_1 = iterator_3_1.value;
            if (date_1.end.isBefore(adapter)) continue;
            return true;
          }
        } catch (e_5_1) {
          e_5 = { error: e_5_1 };
        } finally {
          try {
            if (iterator_3_1 && !iterator_3_1.done && (_a = iterator_3.return)) _a.call(iterator_3);
          } finally {
            if (e_5) throw e_5.error;
          }
        }
        return false;
      }
      try {
        for (
          var _c = __values(this._run({ start: adapter })), _d = _c.next();
          !_d.done;
          _d = _c.next()
        ) {
          var day = _d.value;
          if (options.excludeStart && day.isEqual(adapter)) {
            continue;
          }
          return true;
        }
      } catch (e_6_1) {
        e_6 = { error: e_6_1 };
      } finally {
        try {
          if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
        } finally {
          if (e_6) throw e_6.error;
        }
      }
      return false;
    };
    /**
     * Returns true if an occurrence starts before the provided datetime.
     * If the `excludeStart` option is provided, then occurrences
     * equal to the provided datetime are ignored.
     *
     * If the occurrence generator has a duration, and `excludeStart` is
     * also provided, then this will only return true if an occurrence
     * both starts and ends before the provided datetime.
     */
    OccurrenceGenerator.prototype.occursBefore = function(date, options) {
      var e_7, _a, e_8, _b;
      if (options === void 0) {
        options = {};
      }
      var adapter = this.normalizeDateInput(date);
      if (this.hasDuration && options.excludeStart) {
        try {
          for (
            var _c = __values(this._run({ end: adapter, reverse: true })), _d = _c.next();
            !_d.done;
            _d = _c.next()
          ) {
            var day = _d.value;
            if (day.end.isAfterOrEqual(adapter)) continue;
            return true;
          }
        } catch (e_7_1) {
          e_7 = { error: e_7_1 };
        } finally {
          try {
            if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
          } finally {
            if (e_7) throw e_7.error;
          }
        }
        return false;
      }
      try {
        for (
          var _e = __values(this._run({ end: adapter, reverse: true })), _f = _e.next();
          !_f.done;
          _f = _e.next()
        ) {
          var day = _f.value;
          if (options.excludeStart && day.isEqual(adapter)) {
            continue;
          }
          return true;
        }
      } catch (e_8_1) {
        e_8 = { error: e_8_1 };
      } finally {
        try {
          if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
        } finally {
          if (e_8) throw e_8.error;
        }
      }
      return false;
    };
    Object.defineProperty(OccurrenceGenerator.prototype, 'dateAdapter', {
      get: function() {
        return core.DateAdapterBase.adapter;
      },
      enumerable: false,
      configurable: true,
    });
    OccurrenceGenerator.prototype.normalizeOccurrencesArgs = function(rawArgs) {
      return __assign(__assign({}, rawArgs), {
        start: this.normalizeDateInput(rawArgs.start),
        end: this.normalizeDateInput(rawArgs.end),
      });
    };
    OccurrenceGenerator.prototype.normalizeCollectionsArgs = function(rawArgs) {
      if (rawArgs.reverse !== undefined) {
        throw new core.ArgumentError(
          '`collections()` does not support the `reverse` option at this time.',
        );
      }
      return __assign(__assign({}, rawArgs), {
        start: this.normalizeDateInput(rawArgs.start),
        end: this.normalizeDateInput(rawArgs.end),
      });
    };
    OccurrenceGenerator.prototype.normalizeOccursOnArgs = function(rawArgs) {
      var _this = this;
      if (rawArgs === void 0) {
        rawArgs = {};
      }
      return __assign(__assign({}, rawArgs), {
        date: this.normalizeDateInput(rawArgs.date),
        after: this.normalizeDateInput(rawArgs.after),
        before: this.normalizeDateInput(rawArgs.before),
        excludeDates:
          rawArgs.excludeDates &&
          rawArgs.excludeDates.map(function(date) {
            return _this.normalizeDateInput(date);
          }),
      });
    };
    OccurrenceGenerator.prototype.normalizeRunArgs = function(args) {
      return __assign(__assign({}, args), {
        start: this.normalizeDateInput(args.start),
        end: this.normalizeDateInput(args.end),
      });
    };
    OccurrenceGenerator.prototype.normalizeDateInput = function(date) {
      return date ? core.dateInputToDateTime(date, this.timezone) : undefined;
    };
    OccurrenceGenerator.prototype.normalizeDateInputToAdapter = function(date) {
      if (!date) return;
      return core.dateInputToDateAdapter(date);
    };
    OccurrenceGenerator.prototype.normalizeRunOutput = function(date) {
      return core.normalizeDateTimeTimezone(date, this.timezone);
    };
    OccurrenceGenerator.prototype.getMaxDuration = function(method, options) {
      var maxDuration = options.maxDuration || this.maxDuration;
      if (!Number.isInteger(maxDuration)) {
        throw new core.ArgumentError(
          'When an occurrence generator ' +
            "has a duration, a 'maxDuration' argument must be supplied " +
            ('to ' + method + '().'),
        );
      }
      return maxDuration;
    };
    return OccurrenceGenerator;
  })();
  var OccurrenceIterator = /** @class */ (function() {
    function OccurrenceIterator(iterable, args) {
      var _this = this;
      this.iterable = iterable;
      this.args = args;
      // Need to assert the return type of these methods to prevent typescript from
      // incorrectly reducing them to `DateAdapterBase & { generators: G }`.
      this[Symbol.iterator] = function() {
        return _this.occurrenceIterator();
      };
      this.iterator = iterable._run(args);
      this.isInfinite = iterable.isInfinite;
    }
    OccurrenceIterator.prototype.next = function(args) {
      return this.occurrenceIterator(args).next();
    };
    OccurrenceIterator.prototype.toArray = function() {
      if (this.args.end || this.args.take || !this.isInfinite) {
        return Array.from(this.occurrenceIterator());
      }
      throw new core.InfiniteLoopError(
        'OccurrenceIterator#toArray() can only be called if the iterator ' +
          'is not infinite, or you provide and `end` argument, or you provide ' +
          'a `take` argument.',
      );
    };
    OccurrenceIterator.prototype.occurrenceIterator = function(rawArgs) {
      var args, date, yieldArgs;
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            args = this.normalizeRunArgs(rawArgs);
            date = this.iterator.next(args).value;
            _a.label = 1;
          case 1:
            if (!date) return [3 /*break*/, 3];
            return [4 /*yield*/, this.normalizeDateOutput(date)];
          case 2:
            yieldArgs = _a.sent();
            args = this.normalizeRunArgs(yieldArgs);
            date = this.iterator.next(args).value;
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/, undefined];
        }
      });
    };
    OccurrenceIterator.prototype.normalizeRunArgs = function(args) {
      return {
        skipToDate: this.normalizeDateInput(args && args.skipToDate),
      };
    };
    OccurrenceIterator.prototype.normalizeDateInput = function(date) {
      return date ? core.dateInputToDateTime(date, this.iterable.timezone) : undefined;
    };
    OccurrenceIterator.prototype.normalizeDateOutput = function(date) {
      if (!date) return;
      return date ? core.DateAdapterBase.adapter.fromDateTime(date) : undefined;
    };
    return OccurrenceIterator;
  })();
  var Collection = /** @class */ (function() {
    function Collection(dates, granularity, periodStart, periodEnd) {
      if (dates === void 0) {
        dates = [];
      }
      this.dates = dates;
      this.granularity = granularity;
      this.periodStart = periodStart;
      this.periodEnd = periodEnd;
    }
    return Collection;
  })();
  var CollectionIterator = /** @class */ (function() {
    function CollectionIterator(iterable, args) {
      var _this = this;
      this.iterable = iterable;
      this.args = args;
      this.granularity = 'year';
      this[Symbol.iterator] = function() {
        return _this.iterator;
      };
      if (args.granularity) {
        this.granularity = args.granularity;
        if (this.granularity === 'week' && !args.weekStart) {
          throw new Error('"week" granularity requires `weekStart` arg');
        }
      }
      if (args.weekStart) {
        this.weekStart = args.weekStart;
      }
      if (args.reverse) {
        throw new Error(
          '`OccurrenceGenerator#collections()` does not support iterating in reverse. ' +
            'Though `OccurrenceGenerator#occurrences()` does support iterating in reverse.',
        );
      }
      // Set the end arg, if present, to the end of the period.
      this.args = __assign(__assign({}, args), {
        start: args.start || iterable._run().next().value,
        end: args.end && this.getPeriod(args.end).end,
      });
      this.startDate =
        (this.args.start && this.normalizeDateOutput(this.getPeriod(this.args.start).start)) ||
        null;
      this.iterator = this.collectionIterator();
    }
    CollectionIterator.prototype.next = function() {
      return this.iterator.next();
    };
    /**
     * While `next()` and `[Symbol.iterator]` both share state,
     * `toArray()` does not share state and always returns the whole
     * collections array.
     */
    CollectionIterator.prototype.toArray = function() {
      var e_9, _a;
      if (this.args.end || this.args.take || !this.iterable.isInfinite) {
        var collections = [];
        try {
          for (
            var _b = __values(this.collectionIterator()), _c = _b.next();
            !_c.done;
            _c = _b.next()
          ) {
            var collection = _c.value;
            collections.push(collection);
          }
        } catch (e_9_1) {
          e_9 = { error: e_9_1 };
        } finally {
          try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
          } finally {
            if (e_9) throw e_9.error;
          }
        }
        return collections;
      }
      throw new core.InfiniteLoopError(
        'CollectionIterator#toArray() can only be called if the iterator ' +
          'is not infinite, or you provide and `end` argument, or you provide ' +
          'a `take` argument.',
      );
    };
    CollectionIterator.prototype.normalizeDateOutput = function(date) {
      if (!date) return;
      return core.DateAdapterBase.adapter.fromDateTime(date);
    };
    CollectionIterator.prototype.collectionIterator = function() {
      var iterator, date, period, dates, index;
      var _this = this;
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            if (!this.startDate) return [2 /*return*/];
            iterator = this.occurrenceIterator();
            date = iterator.next().value;
            if (!date) return [2 /*return*/];
            period = this.getPeriod(this.args.start);
            dates = [];
            index = 0;
            _a.label = 1;
          case 1:
            if (!(date && (this.args.take === undefined || this.args.take > index)))
              return [3 /*break*/, 3];
            while (date && date.isBeforeOrEqual(period.end)) {
              dates.push(date);
              date = iterator.next().value;
            }
            return [
              4 /*yield*/,
              new Collection(
                dates.map(function(date) {
                  return _this.normalizeDateOutput(date);
                }),
                this.granularity,
                this.normalizeDateOutput(period.start),
                this.normalizeDateOutput(period.end),
              ),
            ];
          case 2:
            _a.sent();
            if (!date) return [2 /*return*/];
            dates = [];
            period = !this.args.skipEmptyPeriods
              ? this.getPeriod(this.incrementPeriod(period.period))
              : this.getPeriod(date);
            // With these args, periods may overlap and the same date may show up
            // in two periods. Because of this, we need to reset the iterator
            // (otherwise it won't return a date it has already returned).
            if (this.granularity === 'month' && this.weekStart) {
              iterator = this.iterable._run({
                start: period.start,
                end: this.args.end,
              });
              date = iterator.next().value;
            }
            index++;
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/];
        }
      });
    };
    CollectionIterator.prototype.getPeriod = function(date) {
      var start;
      var end;
      var period;
      if (this.granularity === 'month' && this.weekStart) {
        start = date.granularity('month').granularity('week', { weekStart: this.weekStart });
        end = date.endGranularity('month').endGranularity('week', { weekStart: this.weekStart });
        period = date.granularity('month');
      } else {
        start = date.granularity(this.granularity, { weekStart: this.weekStart });
        end = date.endGranularity(this.granularity, { weekStart: this.weekStart });
        period = start;
      }
      return { start: start, end: end, period: period };
    };
    CollectionIterator.prototype.incrementPeriod = function(date) {
      return date.add(1, this.granularity);
    };
    CollectionIterator.prototype.occurrenceIterator = function() {
      var start = this.args.start || this.iterable._run().next().value;
      if (!start) return this.iterable._run(this.args);
      start = this.getPeriod(start).start;
      return this.iterable._run({
        start: start,
        end: this.args.end,
      });
    };
    return CollectionIterator;
  })();
  var Operator = /** @class */ (function(_super) {
    __extends(Operator, _super);
    function Operator(streams, config) {
      var _this = _super.call(this, config) || this;
      _this.streams = streams;
      _this.config = config;
      _this.timezone = config.timezone;
      _this.streams = streams.map(function(stream) {
        return stream instanceof Operator ? stream : stream.set('timezone', _this.timezone);
      });
      _this.isInfinite = _this.calculateIsInfinite();
      _this.hasDuration = _this.calculateHasDuration();
      return _this;
    }
    Operator.prototype.normalizeDateInput = function(date) {
      if (!date) return;
      return core.dateInputToDateTime(date, this.timezone);
    };
    Operator.prototype.normalizeRunOutput = function(date) {
      return core.normalizeDateTimeTimezone(date, this.timezone);
    };
    return Operator;
  })(OccurrenceGenerator);

  /**
   * Simple class that lets us access the `value`, `runArgs`, and `done`
   * status of an OccurrenceGenerator iterator
   */
  var IterableWrapper = /** @class */ (function() {
    function IterableWrapper(generator, runArgs) {
      this.runArgs = runArgs;
      this.stream = generator._run(this.runArgs);
      this.next();
    }
    IterableWrapper.prototype.next = function(args) {
      var _a = this.stream.next(args),
        done = _a.done,
        value = _a.value;
      this.done = typeof done === 'boolean' ? done : true;
      this.value = value;
      return { done: done, value: value };
    };
    return IterableWrapper;
  })();
  function processYieldArgs(streams, options, yieldArgs) {
    var e_1, _a;
    if (options === void 0) {
      options = {};
    }
    if (yieldArgs === void 0) {
      yieldArgs = {};
    }
    if (!yieldArgs.skipToDate || streams.length === 0) return;
    // check for invalid `skipToDate` option
    if (
      options.reverse
        ? streams.every(function(s) {
            return s.value.isBeforeOrEqual(yieldArgs.skipToDate);
          })
        : streams.every(function(s) {
            return s.value.isAfterOrEqual(yieldArgs.skipToDate);
          })
    ) {
      throw new Error(
        'A provided `skipToDate` option must be greater than the last yielded date ' +
          '(or smaller, in the case of reverse iteration)',
      );
    }
    try {
      for (
        var streams_1 = __values(streams), streams_1_1 = streams_1.next();
        !streams_1_1.done;
        streams_1_1 = streams_1.next()
      ) {
        var stream = streams_1_1.value;
        if (stream.done) continue; // no point in calling `next()`
        if (
          options.reverse
            ? stream.value.isBeforeOrEqual(yieldArgs.skipToDate)
            : stream.value.isAfterOrEqual(yieldArgs.skipToDate)
        ) {
          // This can happen there are two streams and one stream starts after the other finishes.
          // Or, when iterating in reverse, when one stream ends before the other starts.
          // In this case we don't want to call `next()` because it will throw an error.
          // In both of these cases, calling `next()` won't do anything anyway.
          continue;
        }
        stream.next(yieldArgs);
      }
    } catch (e_1_1) {
      e_1 = { error: e_1_1 };
    } finally {
      try {
        if (streams_1_1 && !streams_1_1.done && (_a = streams_1.return)) _a.call(streams_1);
      } finally {
        if (e_1) throw e_1.error;
      }
    }
  }
  /** sorts ascending with completed iterables at the end */
  function streamsComparer(a, b) {
    if (a.done && b.done) return 0;
    if (a.done) return 1;
    if (b.done) return -1;
    if (a.value.isAfter(b.value)) return 1;
    return -1;
  }
  /** sorts descending with completed iterables at the start */
  function streamsReverseComparer(a, b) {
    if (a.done && b.done) return 0;
    if (a.done) return -1;
    if (b.done) return 1;
    if (a.value.isAfter(b.value)) return -1;
    return 1;
  }
  function selectNextIterable(streams, options, yieldArgs) {
    if (options === void 0) {
      options = {};
    }
    if (yieldArgs === void 0) {
      yieldArgs = {};
    }
    processYieldArgs(streams, options, yieldArgs);
    return streams
      .sort(options.reverse ? streamsReverseComparer : streamsComparer)
      .filter(function(s) {
        return !s.done;
      })
      .shift();
  }
  function selectLastIterable(streams, options, yieldArgs) {
    if (options === void 0) {
      options = {};
    }
    if (yieldArgs === void 0) {
      yieldArgs = {};
    }
    processYieldArgs(streams, options, yieldArgs);
    return streams
      .sort(options.reverse ? streamsReverseComparer : streamsComparer)
      .filter(function(s) {
        return !s.done;
      })
      .pop();
  }

  /**
   * An operator function which accepts a spread of occurrence generators
   * and adds their occurrences to the output.
   *
   * @param streams a spread of occurrence generators
   */
  function add() {
    var streams = [];
    for (var _i = 0; _i < arguments.length; _i++) {
      streams[_i] = arguments[_i];
    }
    return function(options) {
      return new AddOperator(streams, options);
    };
  }
  var AddOperator = /** @class */ (function(_super) {
    __extends(AddOperator, _super);
    function AddOperator() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    AddOperator.prototype.set = function(_, value) {
      return new AddOperator(
        this.streams.map(function(stream) {
          return stream.set('timezone', value);
        }),
        __assign(__assign({}, this.config), {
          base: this.config.base && this.config.base.set('timezone', value),
          timezone: value,
        }),
      );
    };
    AddOperator.prototype._run = function(args) {
      var wrappedStreams, stream, yieldArgs;
      if (args === void 0) {
        args = {};
      }
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            wrappedStreams = this.streams.map(function(input) {
              return new IterableWrapper(input, args);
            });
            if (this.config.base) {
              wrappedStreams.push(new IterableWrapper(this.config.base, args));
            }
            if (wrappedStreams.length === 0) return [2 /*return*/];
            stream = selectNextIterable(wrappedStreams, args);
            _a.label = 1;
          case 1:
            if (!(stream && !stream.done)) return [3 /*break*/, 3];
            return [4 /*yield*/, this.normalizeRunOutput(stream.value)];
          case 2:
            yieldArgs = _a.sent();
            if (!(yieldArgs && yieldArgs.skipToDate)) {
              // iterate the current stream
              stream.next();
            }
            // select the next stream
            stream = selectNextIterable(wrappedStreams, args, yieldArgs);
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/];
        }
      });
    };
    AddOperator.prototype.calculateIsInfinite = function() {
      return (
        (this.config.base && this.config.base.isInfinite) ||
        this.streams.some(function(stream) {
          return stream.isInfinite;
        })
      );
    };
    AddOperator.prototype.calculateHasDuration = function() {
      var streamsDuration = this.streams.every(function(stream) {
        return stream.hasDuration;
      });
      if (!this.config.base) return streamsDuration;
      return this.config.base.hasDuration && streamsDuration;
    };
    return AddOperator;
  })(Operator);

  var Calendar = /** @class */ (function(_super) {
    __extends(Calendar, _super);
    /**
     *
     * Create a new Calendar object with the specified options.
     *
     * ### Options
     *
     * - **timezone**: The timezone that yielded occurrences should be *displayed* in.
     *   Note, this one affects the *displayed* timezone of yielded occurrences.
     *   For rules, occurrences are first found using the unmodified rule
     *   config (including whatever timezone the `start` datetime is defined
     *   in), and then converted to the timezone specified here before being
     *   yielded. By default, the timezone is *local* time (`null`). So if you don't
     *   want your rules to be displayed in local time, you must supply a
     *   timezone argument.
     * - **data**: arbitrary data you can associate with this Calendar. This
     *   is the only mutable property of `Calendar` objects.
     * - **maxDuration**: currently unused.
     * - **schedules**: either an occurrence generator, or an array of occurrence generators, which should
     *   be used to build up this Calendar's occurrences. The Calendar will display the union of occurrences
     *   its schedules produce.
     *
     */
    function Calendar(args) {
      if (args === void 0) {
        args = {};
      }
      var _this = _super.call(this, args) || this;
      _this.schedules = [];
      _this.data = args.data;
      if (args.schedules) {
        _this.schedules = Array.isArray(args.schedules) ? args.schedules : [args.schedules];
        _this.schedules = _this.schedules.map(function(schedule) {
          return schedule.set('timezone', _this.timezone);
        });
      }
      _this.isInfinite = _this.schedules.some(function(schedule) {
        return schedule.isInfinite;
      });
      _this.hasDuration = _this.schedules.every(function(schedule) {
        return schedule.hasDuration;
      });
      return _this;
    }
    Calendar.prototype.occurrences = function(args) {
      if (args === void 0) {
        args = {};
      }
      return new OccurrenceIterator(this, this.normalizeOccurrencesArgs(args));
    };
    Calendar.prototype.collections = function(args) {
      if (args === void 0) {
        args = {};
      }
      return new CollectionIterator(this, this.normalizeCollectionsArgs(args));
    };
    Calendar.prototype.set = function(prop, value, options) {
      if (prop === 'timezone') {
        return new Calendar({
          schedules: this.schedules.map(function(schedule) {
            return schedule.set(prop, value, options);
          }),
          data: this.data,
          timezone: value,
          maxDuration: this.maxDuration,
        });
      } else if (prop === 'schedules') {
        return new Calendar({
          schedules: Array.isArray(value) ? value : [value],
          data: this.data,
          timezone: this.timezone,
          maxDuration: this.maxDuration,
        });
      }
      throw new core.ArgumentError('Unknown value for `prop`: ' + ('"' + prop + '"'));
    };
    Calendar.prototype._run = function(args) {
      var count, iterator, date, index, yieldArgs;
      if (args === void 0) {
        args = {};
      }
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            if (this.schedules.length === 0) return [2 /*return*/];
            count = args.take;
            delete args.take;
            if (this.schedules.length === 1) {
              iterator = this.schedules[0]._run(args);
            } else {
              iterator = new AddOperator(this.schedules, {
                timezone: this.timezone,
              })._run(args);
            }
            date = iterator.next().value;
            index = 0;
            _a.label = 1;
          case 1:
            if (!(date && (count === undefined || count > index))) return [3 /*break*/, 3];
            date = date.add(this, 'generator');
            return [4 /*yield*/, this.normalizeRunOutput(date)];
          case 2:
            yieldArgs = _a.sent();
            date = iterator.next(yieldArgs).value;
            index++;
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/, undefined];
        }
      });
    };
    return Calendar;
  })(OccurrenceGenerator);

  var Dates = /** @class */ (function(_super) {
    __extends(Dates, _super);
    function Dates(args) {
      if (args === void 0) {
        args = {};
      }
      var _this = _super.call(this, args) || this;
      _this.adapters = [];
      _this.isInfinite = false;
      _this.maxDuration = 0;
      _this.datetimes = [];
      _this.data = args.data;
      if (args.dates) {
        _this.adapters = args.dates.map(function(date) {
          var adapter = _this.normalizeDateInputToAdapter(date);
          if (args.duration && !adapter.duration) {
            adapter = adapter.set('duration', args.duration);
          }
          return adapter.set('timezone', _this.timezone);
        });
        _this.datetimes = _this.adapters.map(function(adapter) {
          return adapter.toDateTime();
        });
      }
      _this.hasDuration = _this.datetimes.every(function(date) {
        return !!date.duration;
      });
      if (_this.hasDuration) {
        _this.maxDuration = _this.adapters.reduce(function(prev, curr) {
          return curr.duration > prev ? curr.duration : prev;
        }, 0);
      }
      return _this;
    }
    Object.defineProperty(Dates.prototype, 'length', {
      get: function() {
        return this.adapters.length;
      },
      enumerable: false,
      configurable: true,
    });
    Object.defineProperty(Dates.prototype, 'firstDate', {
      /** Returns the first occurrence or, if there are no occurrences, null. */
      get: function() {
        return this.adapters[0] || null;
      },
      enumerable: false,
      configurable: true,
    });
    Object.defineProperty(Dates.prototype, 'lastDate', {
      /** Returns the last occurrence or, if there are no occurrences, null. */
      get: function() {
        return this.adapters[this.length - 1] || null;
      },
      enumerable: false,
      configurable: true,
    });
    Dates.prototype.occurrences = function(args) {
      if (args === void 0) {
        args = {};
      }
      return new OccurrenceIterator(this, this.normalizeOccurrencesArgs(args));
    };
    Dates.prototype.collections = function(args) {
      if (args === void 0) {
        args = {};
      }
      return new CollectionIterator(this, this.normalizeCollectionsArgs(args));
    };
    Dates.prototype.add = function(value) {
      return new Dates({
        dates: __spread(this.adapters, [value]),
        timezone: this.timezone,
        data: this.data,
      });
    };
    Dates.prototype.remove = function(value) {
      var dates = this.adapters.slice();
      var input = this.normalizeDateInputToAdapter(value);
      var index = dates.findIndex(function(date) {
        return date.valueOf() === input.valueOf();
      });
      if (index >= 0) {
        dates.splice(index, 1);
      }
      return new Dates({
        dates: dates,
        timezone: this.timezone,
        data: this.data,
      });
    };
    Dates.prototype.set = function(prop, value, options) {
      var _this = this;
      if (options === void 0) {
        options = {};
      }
      var timezone = this.timezone;
      var dates = this.adapters.slice();
      if (prop === 'timezone') {
        if (value === this.timezone) return this;
        else if (options.keepLocalTime) {
          dates = this.adapters.map(function(adapter) {
            var json = adapter.toJSON();
            json.timezone = value;
            return _this.dateAdapter.fromJSON(json);
          });
        }
        timezone = value;
      } else if (prop === 'dates') {
        dates = value;
      } else if (prop === 'duration') {
        dates = dates.map(function(date) {
          return date.set('duration', value || 0);
        });
      } else {
        throw new core.ArgumentError(
          'Unexpected prop argument "' + prop + '". Accepted values are "timezone" or "dates"',
        );
      }
      return new Dates({
        dates: dates,
        data: this.data,
        timezone: timezone,
      });
    };
    Dates.prototype.filter = function(fn) {
      return new Dates({
        dates: this.adapters.filter(fn),
        data: this.data,
        timezone: this.timezone,
      });
    };
    Dates.prototype._run = function(args) {
      var dates, dateCache, date, yieldArgs;
      if (args === void 0) {
        args = {};
      }
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            dates = this.datetimes.sort(core.dateTimeSortComparer);
            if (args.start) {
              dates = dates.filter(function(date) {
                return date.isAfterOrEqual(args.start);
              });
            }
            if (args.end) {
              dates = dates.filter(function(date) {
                return date.isBeforeOrEqual(args.end);
              });
            }
            if (args.reverse) {
              dates = dates.slice().reverse();
            }
            if (args.take) {
              dates = dates.slice(0, args.take);
            }
            dateCache = dates.slice();
            date = dateCache.shift();
            _a.label = 1;
          case 1:
            if (!date) return [3 /*break*/, 3];
            if (
              yieldArgs &&
              yieldArgs.skipToDate &&
              (args.reverse
                ? yieldArgs.skipToDate.isBefore(date)
                : yieldArgs.skipToDate.isAfter(date))
            ) {
              date = dateCache.shift();
              return [3 /*break*/, 1];
            }
            date = date.add(this, 'generator');
            return [4 /*yield*/, this.normalizeRunOutput(date)];
          case 2:
            yieldArgs = _a.sent();
            // Here, we are matching the behavior of the RecurrenceRulesIterator
            if (
              yieldArgs &&
              yieldArgs.skipToDate &&
              (args.reverse
                ? yieldArgs.skipToDate.isAfterOrEqual(date)
                : yieldArgs.skipToDate.isBeforeOrEqual(date))
            ) {
              throw new Error(
                'A provided `skipToDate` option must be greater than the last yielded date ' +
                  '(or smaller, in the case of reverse iteration)',
              );
            }
            date = dateCache.shift();
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/, undefined];
        }
      });
    };
    return Dates;
  })(OccurrenceGenerator);

  var InvalidDateAdapterError = /** @class */ (function(_super) {
    __extends(InvalidDateAdapterError, _super);
    function InvalidDateAdapterError() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    return InvalidDateAdapterError;
  })(Error);
  var dateAdapterConfig;
  var DateAdapterBase = /** @class */ (function() {
    function DateAdapterBase(_date, options) {
      /**
       * The metadata property is intended to store arbitrary data
       * related to this DateAdapter. Unlike other properties, the
       * metadata property is intended to be mutable.
       */
      this.metadata = {};
      this.duration = (options && options.duration) || 0;
      this.generators = (options && options.generators && options.generators.slice()) || [];
      if (!Number.isInteger(this.duration) || this.duration < 0) {
        throw new InvalidDateAdapterError('duration must be a non-negative integer');
      }
    }
    Object.defineProperty(DateAdapterBase, 'adapter', {
      get: function() {
        if (!dateAdapterConfig) {
          throw new Error('No date adapter has been configured. See rSchedule docs.');
        }
        return dateAdapterConfig;
      },
      set: function(value) {
        if (dateAdapterConfig) {
          throw new Error('"' + dateAdapterConfig.name + '" has already been configured.');
        }
        dateAdapterConfig = value;
      },
      enumerable: false,
      configurable: true,
    });
    DateAdapterBase.isDate = function(_object) {
      throw unimplementedError('isDate()');
    };
    DateAdapterBase.fromDate = function(_date, _options) {
      throw unimplementedError('fromDate()');
    };
    DateAdapterBase.fromJSON = function(_json) {
      throw unimplementedError('fromJSON()');
    };
    DateAdapterBase.fromDateTime = function(_datetime) {
      throw unimplementedError('fromDateTime()');
    };
    /**
     * Returns a string in simplified extended ISO format (ISO 8601).
     *
     * _Note: this method is intended for testing and its
     * implementation isn't particularly performant._
     */
    DateAdapterBase.prototype.toISOString = function() {
      return this.set('timezone', 'UTC')
        .toDateTime()
        .toISOString();
    };
    DateAdapterBase.prototype.toDateTime = function() {
      var date = DateTime.fromJSON(
        __assign(__assign({}, this.toJSON()), {
          generators: this.generators,
          metadata: this.metadata,
        }),
      );
      return date;
    };
    DateAdapterBase.hasTimezoneSupport = false;
    return DateAdapterBase;
  })();
  function unimplementedError(name) {
    return new Error('You must implement the "' + name + '" method for this DateAdapter class');
  }
  var DateAdapter;
  (function(DateAdapter) {
    DateAdapter.WEEKDAYS = ['SU', 'MO', 'TU', 'WE', 'TH', 'FR', 'SA'];
    DateAdapter.MILLISECONDS_IN_SECOND = 1000;
    DateAdapter.MILLISECONDS_IN_MINUTE = DateAdapter.MILLISECONDS_IN_SECOND * 60;
    DateAdapter.MILLISECONDS_IN_HOUR = DateAdapter.MILLISECONDS_IN_MINUTE * 60;
    DateAdapter.MILLISECONDS_IN_DAY = DateAdapter.MILLISECONDS_IN_HOUR * 24;
    DateAdapter.MILLISECONDS_IN_WEEK = DateAdapter.MILLISECONDS_IN_DAY * 7;
  })(DateAdapter || (DateAdapter = {}));
  var InvalidDateTimeError = /** @class */ (function(_super) {
    __extends(InvalidDateTimeError, _super);
    function InvalidDateTimeError() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    return InvalidDateTimeError;
  })(Error);
  var DateTime = /** @class */ (function() {
    function DateTime(date, timezone, duration, generators, options) {
      if (options === void 0) {
        options = {};
      }
      this.metadata = {};
      this.date = new Date(date);
      this.timezone = timezone || null;
      this.duration = duration || 0;
      this.generators = (generators && generators.slice()) || [];
      if (!Number.isInteger(this.duration) || this.duration < 0) {
        throw new InvalidDateTimeError('duration must be a non-negative integer');
      }
      if (options.metadata) {
        Object.assign(this.metadata, options.metadata);
      }
      this.assertIsValid();
    }
    // /**
    //  * Similar to `Array.isArray()`, `isInstance()` provides a surefire method
    //  * of determining if an object is a `DateTime` by checking against the
    //  * global symbol registry.
    //  */
    // static isInstance(object: any): object is DateTime {
    //   return !!(object && object[DATETIME_ID]);
    // }
    DateTime.fromJSON = function(json) {
      var date = new Date(
        Date.UTC(
          json.year,
          json.month - 1,
          json.day,
          json.hour,
          json.minute,
          json.second,
          json.millisecond,
        ),
      );
      return new DateTime(date, json.timezone, json.duration, json.generators, {
        metadata: json.metadata,
      });
    };
    DateTime.fromDateAdapter = function(adapter) {
      return DateTime.fromJSON(
        __assign(__assign({}, adapter.toJSON()), {
          generators: adapter.generators,
          metadata: adapter.metadata,
        }),
      );
    };
    Object.defineProperty(DateTime.prototype, 'end', {
      /**
       * Returns `undefined` if `duration` is `0`. Else returns
       * the `end` date.
       */
      get: function() {
        if (!this.duration) return;
        if (this._end) return this._end;
        this._end = this.add(this.duration, 'millisecond');
        return this._end;
      },
      enumerable: false,
      configurable: true,
    });
    // While we constrain the argument to be another DateAdapter in typescript
    // we handle the case of someone passing in another type of object in javascript
    DateTime.prototype.isEqual = function(object) {
      if (!object) {
        return false;
      }
      assertSameTimeZone(this, object);
      return this.valueOf() === object.valueOf();
    };
    DateTime.prototype.isBefore = function(object) {
      assertSameTimeZone(this, object);
      return this.valueOf() < object.valueOf();
    };
    DateTime.prototype.isBeforeOrEqual = function(object) {
      assertSameTimeZone(this, object);
      return this.valueOf() <= object.valueOf();
    };
    DateTime.prototype.isAfter = function(object) {
      assertSameTimeZone(this, object);
      return this.valueOf() > object.valueOf();
    };
    DateTime.prototype.isAfterOrEqual = function(object) {
      assertSameTimeZone(this, object);
      return this.valueOf() >= object.valueOf();
    };
    DateTime.prototype.isOccurring = function(object) {
      if (!this.duration) {
        throw new Error('DateTime#isOccurring() is only applicable to DateTimes with durations');
      }
      assertSameTimeZone(this, object);
      return (
        object.isAfterOrEqual(this) &&
        object.isBeforeOrEqual(this.add(this.duration, 'millisecond'))
      );
    };
    DateTime.prototype.add = function(amount, unit) {
      switch (unit) {
        case 'generator': {
          var generators = this.generators.slice();
          generators.unshift(amount);
          return new DateTime(this.date, this.timezone, this.duration, generators);
        }
        case 'year':
          return this.forkDateTime(addUTCYears(this.date, amount));
        case 'month':
          return this.forkDateTime(addUTCMonths(this.date, amount));
        case 'week':
          return this.forkDateTime(addUTCWeeks(this.date, amount));
        case 'day':
          return this.forkDateTime(addUTCDays(this.date, amount));
        case 'hour':
          return this.forkDateTime(addUTCHours(this.date, amount));
        case 'minute':
          return this.forkDateTime(addUTCMinutes(this.date, amount));
        case 'second':
          return this.forkDateTime(addUTCSeconds(this.date, amount));
        case 'millisecond':
          return this.forkDateTime(addUTCMilliseconds(this.date, amount));
        default:
          throw new Error('Invalid unit provided to `DateTime#add`');
      }
    };
    DateTime.prototype.subtract = function(amount, unit) {
      switch (unit) {
        case 'year':
          return this.forkDateTime(subUTCYears(this.date, amount));
        case 'month':
          return this.forkDateTime(subUTCMonths(this.date, amount));
        case 'week':
          return this.forkDateTime(subUTCWeeks(this.date, amount));
        case 'day':
          return this.forkDateTime(subUTCDays(this.date, amount));
        case 'hour':
          return this.forkDateTime(subUTCHours(this.date, amount));
        case 'minute':
          return this.forkDateTime(subUTCMinutes(this.date, amount));
        case 'second':
          return this.forkDateTime(subUTCSeconds(this.date, amount));
        case 'millisecond':
          return this.forkDateTime(subUTCMilliseconds(this.date, amount));
        default:
          throw new Error('Invalid unit provided to `DateTime#subtract`');
      }
    };
    DateTime.prototype.get = function(unit) {
      switch (unit) {
        case 'year':
          return this.date.getUTCFullYear();
        case 'month':
          return this.date.getUTCMonth() + 1;
        case 'yearday':
          return getUTCYearDay(this.date);
        case 'weekday':
          return DateAdapter.WEEKDAYS[this.date.getUTCDay()];
        case 'day':
          return this.date.getUTCDate();
        case 'hour':
          return this.date.getUTCHours();
        case 'minute':
          return this.date.getUTCMinutes();
        case 'second':
          return this.date.getUTCSeconds();
        case 'millisecond':
          return this.date.getUTCMilliseconds();
        default:
          throw new Error('Invalid unit provided to `DateTime#set`');
      }
    };
    DateTime.prototype.set = function(unit, value) {
      if (unit === 'duration') {
        return new DateTime(this.date, this.timezone, value, this.generators);
      }
      if (unit === 'generators') {
        return new DateTime(this.date, this.timezone, this.duration, value);
      }
      var date = new Date(this.date);
      switch (unit) {
        case 'year':
          date.setUTCFullYear(value);
          break;
        case 'month': {
          // If the current day of the month
          // is greater than days in the month we are moving to, we need to also
          // set the day to the end of that month.
          var length_1 = monthLength(value, date.getUTCFullYear());
          var day = date.getUTCDate();
          if (day > length_1) {
            date.setUTCDate(1);
            date.setUTCMonth(value);
            date = subUTCDays(date, 1);
          } else {
            date.setUTCMonth(value - 1);
          }
          break;
        }
        case 'day':
          date.setUTCDate(value);
          break;
        case 'hour':
          date.setUTCHours(value);
          break;
        case 'minute':
          date.setUTCMinutes(value);
          break;
        case 'second':
          date.setUTCSeconds(value);
          break;
        case 'millisecond':
          date.setUTCMilliseconds(value);
          break;
        default:
          throw new Error('Invalid unit provided to `DateTime#set`');
      }
      return this.forkDateTime(date);
    };
    DateTime.prototype.granularity = function(granularity, opt) {
      if (opt === void 0) {
        opt = {};
      }
      var date = this.forkDateTime(this.date);
      switch (granularity) {
        case 'year':
          date = date.set('month', 1);
        case 'month':
          date = date.set('day', 1);
          break;
        case 'week':
          date = setDateToStartOfWeek(date, opt.weekStart);
      }
      switch (granularity) {
        case 'year':
        case 'month':
        case 'week':
        case 'day':
          date = date.set('hour', 0);
        case 'hour':
          date = date.set('minute', 0);
        case 'minute':
          date = date.set('second', 0);
        case 'second':
          date = date.set('millisecond', 0);
        case 'millisecond':
          return date;
        default:
          throw new Error('Invalid granularity provided to `DateTime#granularity`: ' + granularity);
      }
    };
    DateTime.prototype.endGranularity = function(granularity, opt) {
      if (opt === void 0) {
        opt = {};
      }
      var date = this.forkDateTime(this.date);
      switch (granularity) {
        case 'year':
          date = date.set('month', 12);
        case 'month':
          date = date.set('day', monthLength(date.get('month'), date.get('year')));
          break;
        case 'week':
          date = setDateToEndOfWeek(date, opt.weekStart);
      }
      switch (granularity) {
        case 'year':
        case 'month':
        case 'week':
        case 'day':
          date = date.set('hour', 23);
        case 'hour':
          date = date.set('minute', 59);
        case 'minute':
          date = date.set('second', 59);
        case 'second':
          date = date.set('millisecond', 999);
        case 'millisecond':
          return date;
        default:
          throw new Error('Invalid granularity provided to `DateTime#granularity`: ' + granularity);
      }
    };
    DateTime.prototype.toISOString = function() {
      return this.date.toISOString();
    };
    DateTime.prototype.toDateTime = function() {
      return this;
    };
    DateTime.prototype.toJSON = function() {
      var json = {
        timezone: this.timezone,
        year: this.get('year'),
        month: this.get('month'),
        day: this.get('day'),
        hour: this.get('hour'),
        minute: this.get('minute'),
        second: this.get('second'),
        millisecond: this.get('millisecond'),
      };
      if (this.duration) {
        json.duration = this.duration;
      }
      return json;
    };
    DateTime.prototype.valueOf = function() {
      return this.date.valueOf();
    };
    DateTime.prototype.assertIsValid = function() {
      if (isNaN(this.valueOf())) {
        throw new InvalidDateTimeError('DateTime has invalid date.');
      }
      return true;
    };
    DateTime.prototype.forkDateTime = function(date) {
      return new DateTime(date, this.timezone, this.duration, this.generators);
    };
    return DateTime;
  })();
  function assertSameTimeZone(x, y) {
    if (x.timezone !== y.timezone) {
      throw new InvalidDateTimeError(
        'Attempted to compare a datetime to another date in a different timezone: ' +
          JSON.stringify(x) +
          ' and ' +
          JSON.stringify(y),
      );
    }
    return true;
  }
  function setDateToStartOfWeek(date, wkst) {
    var index = orderedWeekdays(wkst).indexOf(date.get('weekday'));
    return date.subtract(index, 'day');
  }
  function setDateToEndOfWeek(date, wkst) {
    var index = orderedWeekdays(wkst).indexOf(date.get('weekday'));
    return date.add(6 - index, 'day');
  }
  function orderedWeekdays(wkst) {
    if (wkst === void 0) {
      wkst = 'SU';
    }
    var wkdays = DateAdapter.WEEKDAYS.slice();
    var index = wkdays.indexOf(wkst);
    while (index !== 0) {
      shiftArray(wkdays);
      index--;
    }
    return wkdays;
  }
  function shiftArray(array, from) {
    if (from === void 0) {
      from = 'first';
    }
    if (array.length === 0) {
      return array;
    } else if (from === 'first') {
      array.push(array.shift());
    } else {
      array.unshift(array.pop());
    }
    return array;
  }
  /**
   * Returns the days in the given month.
   *
   * @param month base-1
   * @param year
   */
  function monthLength(month, year) {
    var block = {
      1: 31,
      2: getDaysInFebruary(year),
      3: 31,
      4: 30,
      5: 31,
      6: 30,
      7: 31,
      8: 31,
      9: 30,
      10: 31,
      11: 30,
      12: 31,
    };
    return block[month];
  }
  function getDaysInFebruary(year) {
    return isLeapYear(year) ? 29 : 28;
  }
  // taken from date-fn
  function isLeapYear(year) {
    return year % 400 === 0 || (year % 4 === 0 && year % 100 !== 0);
  }
  function getUTCYearDay(now) {
    var start = new Date(Date.UTC(now.getUTCFullYear(), 0, 1));
    var diff = now.valueOf() - start.valueOf();
    return 1 + Math.floor(diff / DateAdapter.MILLISECONDS_IN_DAY);
  }
  /**
   * These functions are basically lifted from `date-fns`, but changed
   * to use the UTC date methods, which `date-fns` doesn't support.
   */
  function toInteger(input) {
    if (input === null || input === true || input === false) {
      return NaN;
    }
    var int = Number(input);
    if (isNaN(int)) {
      return int;
    }
    return int < 0 ? Math.ceil(int) : Math.floor(int);
  }
  function addMilliseconds(dirtyDate, dirtyAmount) {
    if (arguments.length < 2) {
      throw new TypeError('2 arguments required, but only ' + arguments.length + ' present');
    }
    var timestamp = dirtyDate.valueOf();
    var amount = toInteger(dirtyAmount);
    return new Date(timestamp + amount);
  }
  function addUTCYears(date, input) {
    var amount = toInteger(input);
    return addUTCMonths(date, amount * 12);
  }
  function addUTCMonths(date, input) {
    var amount = toInteger(input);
    date = new Date(date);
    var desiredMonth = date.getUTCMonth() + amount;
    var dateWithDesiredMonth = new Date(0);
    dateWithDesiredMonth.setUTCFullYear(date.getUTCFullYear(), desiredMonth, 1);
    dateWithDesiredMonth.setUTCHours(0, 0, 0, 0);
    var daysInMonth = monthLength(
      dateWithDesiredMonth.getUTCMonth() + 1,
      dateWithDesiredMonth.getUTCFullYear(),
    );
    // Set the last day of the new month
    // if the original date was the last day of the longer month
    date.setUTCMonth(desiredMonth, Math.min(daysInMonth, date.getUTCDate()));
    return date;
  }
  function addUTCWeeks(date, input) {
    var amount = toInteger(input);
    var days = amount * 7;
    return addUTCDays(date, days);
  }
  function addUTCDays(date, input) {
    // by adding milliseconds rather than days, we supress the native Date object's automatic
    // daylight savings time conversions which we don't want in UTC mode
    return addUTCMilliseconds(date, toInteger(input) * DateAdapter.MILLISECONDS_IN_DAY);
  }
  function addUTCHours(date, input) {
    var amount = toInteger(input);
    return addMilliseconds(date, amount * DateAdapter.MILLISECONDS_IN_HOUR);
  }
  function addUTCMinutes(date, input) {
    var amount = toInteger(input);
    return addMilliseconds(date, amount * DateAdapter.MILLISECONDS_IN_MINUTE);
  }
  function addUTCSeconds(date, input) {
    var amount = toInteger(input);
    return addMilliseconds(date, amount * DateAdapter.MILLISECONDS_IN_SECOND);
  }
  function addUTCMilliseconds(date, input) {
    var amount = toInteger(input);
    var timestamp = date.getTime();
    return new Date(timestamp + amount);
  }
  function subUTCYears(date, amount) {
    return addUTCYears(date, -amount);
  }
  function subUTCMonths(date, amount) {
    return addUTCMonths(date, -amount);
  }
  function subUTCWeeks(date, amount) {
    return addUTCWeeks(date, -amount);
  }
  function subUTCDays(date, amount) {
    return addUTCDays(date, -amount);
  }
  function subUTCHours(date, amount) {
    return addUTCHours(date, -amount);
  }
  function subUTCMinutes(date, amount) {
    return addUTCMinutes(date, -amount);
  }
  function subUTCSeconds(date, amount) {
    return addUTCSeconds(date, -amount);
  }
  function subUTCMilliseconds(date, amount) {
    return addUTCMilliseconds(date, -amount);
  }

  var ArgumentError = /** @class */ (function(_super) {
    __extends(ArgumentError, _super);
    function ArgumentError() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    return ArgumentError;
  })(Error);
  var InfiniteLoopError = /** @class */ (function(_super) {
    __extends(InfiniteLoopError, _super);
    function InfiniteLoopError() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    return InfiniteLoopError;
  })(Error);
  // export function normalizeDateInput<T extends DateAdapter>(
  //   dateAdapter: DateAdapterConstructor<T>,
  //   input: DateInput<T>,
  //   timezone: string | null,
  // ): DateTime {
  //   if (input instanceof DateTime) {
  //     if (input.timezone !== timezone) {
  //       return dateAdapter
  //         .fromDateTime(input)
  //         .set('timezone', timezone)
  //         .toDateTime();
  //     }
  //     return input;
  //   }
  //   return input instanceof DateAdapter
  //     ? input.set('timezone', timezone).toDateTime()
  //     : new dateAdapter(input).toDateTime();
  // }
  function normalizeDateTimeTimezone(date, timezone) {
    if (date.timezone !== timezone) {
      return DateAdapterBase.adapter
        .fromDateTime(date)
        .set('timezone', timezone)
        .toDateTime();
    }
    return date;
  }

  var RuleBase = /** @class */ (function(_super) {
    __extends(RuleBase, _super);
    function RuleBase(recurrenceRules, config, options) {
      if (options === void 0) {
        options = {};
      }
      var _this = _super.call(this, options) || this;
      _this.recurrenceRules = recurrenceRules;
      _this.options = core.cloneRuleOptions(config);
      _this.normOptions = core.normalizeRuleOptions(_this.recurrenceRules, _this.options);
      _this.timezone =
        options.timezone !== undefined ? options.timezone : _this.normOptions.start.timezone;
      _this.data = options.data;
      _this.hasDuration = !!config.duration;
      if (_this.hasDuration) _this.duration = config.duration;
      _this.isInfinite =
        _this.normOptions.end === undefined && _this.normOptions.count === undefined;
      return _this;
    }
    RuleBase.prototype.occurrences = function(args) {
      if (args === void 0) {
        args = {};
      }
      return new OccurrenceIterator(this, this.normalizeOccurrencesArgs(args));
    };
    RuleBase.prototype.collections = function(args) {
      if (args === void 0) {
        args = {};
      }
      return new CollectionIterator(this, this.normalizeCollectionsArgs(args));
    };
    RuleBase.prototype._run = function(rawArgs) {
      var args, iterator, date, index, yieldArgs;
      if (rawArgs === void 0) {
        rawArgs = {};
      }
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            args = this.normalizeRunArgs(rawArgs);
            iterator = new core.RecurrenceRulesIterator(
              core.recurrenceRulesReducer(this.recurrenceRules),
              this.normOptions,
              args,
            );
            date = iterator.next().value;
            index = 0;
            _a.label = 1;
          case 1:
            if (!(date && (args.take === undefined || index < args.take))) return [3 /*break*/, 3];
            index++;
            date = date.add(this, 'generator');
            return [4 /*yield*/, this.normalizeRunOutput(date)];
          case 2:
            yieldArgs = _a.sent();
            if (yieldArgs === null || yieldArgs === void 0 ? void 0 : yieldArgs.skipToDate) {
              // The RecurrenceRuleIterator might have a different timezone from the rule.
              // Because of this, the yieldArgs will not properly be normalized by the
              // OccurrenceIterator, so we need to do it here
              date = iterator.next(
                __assign(__assign({}, yieldArgs), {
                  skipToDate: normalizeDateTimeTimezone(
                    yieldArgs.skipToDate,
                    iterator.start.timezone,
                  ),
                }),
              ).value;
            } else {
              // theoretically, the yieldArgs are undefined here
              // maybe in the future there will be other yieldArg options though
              date = iterator.next(yieldArgs).value;
            }
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/, undefined];
        }
      });
    };
    return RuleBase;
  })(OccurrenceGenerator);

  var Rule = /** @class */ (function(_super) {
    __extends(Rule, _super);
    /**
     * Create a new Rule object with the specified rule config and options.
     *
     * ### Options
     *
     * - **timezone**: the timezone that yielded occurrences should be in. Note,
     *   this does not change the rule config. Occurrences are first found using
     *   the unmodified rule config, and then converted to the timezone specified
     *   here before being yielded.
     * - **data**: arbitrary data you can associate with this rule. This
     *   is the only mutable property of `Rule` objects.
     *
     * ### Rule Config
     *
     * - #### frequency
     *
     *   The frequency rule part identifies the type of recurrence rule. Valid values
     *   include `"SECONDLY"`, `"MINUTELY"`, `"HOURLY"`, `"DAILY"`, `"WEEKLY"`,
     *   `"MONTHLY"`, or `"YEARLY"`.
     *
     * - #### start
     *
     *   The start of the rule (not necessarily the first occurrence).
     *   Either a `DateAdapter` instance, date object, or `DateTime` object.
     *   The type of date object depends on the `DateAdapter` class used for this
     *   `Rule`.
     *
     * - #### end?
     *
     *   The end of the rule (not necessarily the last occurrence).
     *   Either a `DateAdapter` instance, date object, or `DateTime` object.
     *   The type of date object depends on the `DateAdapter` class used for this
     *   `Rule`.
     *
     * - #### duration?
     *
     *   A length of time expressed in milliseconds.
     *
     * - #### interval?
     *
     *   The interval rule part contains a positive integer representing at
     *   which intervals the recurrence rule repeats. The default value is
     *   `1`, meaning every second for a SECONDLY rule, every minute for a
     *   MINUTELY rule, every hour for an HOURLY rule, every day for a
     *   DAILY rule, every week for a WEEKLY rule, every month for a
     *   MONTHLY rule, and every year for a YEARLY rule. For example,
     *   within a DAILY rule, a value of `8` means every eight days.
     *
     * - #### count?
     *
     *   The count rule part defines the number of occurrences at which to
     *   range-bound the recurrence. `count` and `end` are both two different
     *   ways of specifying how a recurrence completes.
     *
     * - #### weekStart?
     *
     *   The weekStart rule part specifies the day on which the workweek starts.
     *   Valid values are `"MO"`, `"TU"`, `"WE"`, `"TH"`, `"FR"`, `"SA"`, and `"SU"`.
     *   This is significant when a WEEKLY rule has an interval greater than 1,
     *   and a `byDayOfWeek` rule part is specified. The
     *   default value is `"MO"`.
     *
     * - #### bySecondOfMinute?
     *
     *   The bySecondOfMinute rule part expects an array of seconds
     *   within a minute. Valid values are 0 to 60.
     *
     * - #### byMinuteOfHour?
     *
     *   The byMinuteOfHour rule part expects an array of minutes within an hour.
     *   Valid values are 0 to 59.
     *
     * - #### byHourOfDay?
     *
     *   The byHourOfDay rule part expects an array of hours of the day.
     *   Valid values are 0 to 23.
     *
     * - #### byDayOfWeek?
     *
     *   *note: the byDayOfWeek rule part is kinda complex. Blame the ICAL spec.*
     *
     *   The byDayOfWeek rule part expects an array. Each array entry can
     *   be a day of the week (`"SU"`, `"MO"` , `"TU"`, `"WE"`, `"TH"`,
     *   `"FR"`, `"SA"`). If the rule's `frequency` is either MONTHLY or YEARLY,
     *   Any entry can also be a tuple where the first value of the tuple is a
     *   day of the week and the second value is an positive/negative integer
     *   (e.g. `["SU", 1]`). In this case, the number indicates the nth occurrence of
     *   the specified day within the MONTHLY or YEARLY rule.
     *
     *   The behavior of byDayOfWeek changes depending on the `frequency`
     *   of the rule.
     *
     *   Within a MONTHLY rule, `["MO", 1]` represents the first Monday
     *   within the month, whereas `["MO", -1]` represents the last Monday
     *   of the month.
     *
     *   Within a YEARLY rule, the numeric value in a byDayOfWeek tuple entry
     *   corresponds to an offset within the month when the byMonthOfYear rule part is
     *   present, and corresponds to an offset within the year otherwise.
     *
     *   Regardless of rule `frequency`, if a byDayOfWeek entry is a string
     *   (rather than a tuple), it means "all of these days" within the specified
     *   frequency (e.g. within a MONTHLY rule, `"MO"` represents all Mondays within
     *   the month).
     *
     * - #### byDayOfMonth?
     *
     *   The byDayOfMonth rule part expects an array of days
     *   of the month. Valid values are 1 to 31 or -31 to -1.
     *
     *   For example, -10 represents the tenth to the last day of the month.
     *   The byDayOfMonth rule part *must not* be specified when the rule's
     *   `frequency` is set to WEEKLY.
     *
     * - #### byMonthOfYear?
     *
     *   The byMonthOfYear rule part expects an array of months
     *   of the year. Valid values are 1 to 12.
     *
     */
    function Rule(config, options) {
      if (options === void 0) {
        options = {};
      }
      return _super.call(this, Rule.recurrenceRules, config, options) || this;
    }
    Rule.prototype.set = function(prop, value, tzoptions) {
      if (tzoptions === void 0) {
        tzoptions = {};
      }
      var options = core.cloneRuleOptions(this.options);
      var timezone = this.timezone;
      if (prop === 'timezone') {
        if (value === this.timezone && !tzoptions.keepLocalTime) return this;
        else if (tzoptions.keepLocalTime) {
          var json = this.normalizeDateInput(options.start).toJSON();
          json.timezone = value;
          var adapter = this.dateAdapter.fromJSON(json);
          // prettier-ignore
          options.start =
                        options.start instanceof this.dateAdapter ? adapter :
                            options.start instanceof core.DateTime ? adapter.toDateTime() :
                                adapter.date;
        }
        timezone = value;
      } else if (prop === 'options') {
        options = value;
      } else {
        options[prop] = value;
      }
      return new Rule(options, {
        data: this.data,
        maxDuration: this.maxDuration,
        timezone: timezone,
      });
    };
    Rule.recurrenceRules = [];
    return Rule;
  })(RuleBase);

  var ScheduleBase = /** @class */ (function(_super) {
    __extends(ScheduleBase, _super);
    /**
     * Create a new Schedule object with the specified options.
     *
     * The order of precidence for rrules, rdates, exrules, and exdates is:
     *
     * 1. rrules are included
     * 2. exrules are excluded
     * 3. rdates are included
     * 4. exdates are excluded
     *
     */
    function ScheduleBase(options) {
      var _this = _super.call(this, options) || this;
      _this.data = options.data;
      return _this;
    }
    ScheduleBase.prototype.occurrences = function(args) {
      if (args === void 0) {
        args = {};
      }
      return new OccurrenceIterator(this, this.normalizeOccurrencesArgs(args));
    };
    ScheduleBase.prototype.collections = function(args) {
      if (args === void 0) {
        args = {};
      }
      return new CollectionIterator(this, this.normalizeCollectionsArgs(args));
    };
    ScheduleBase.prototype._run = function(args) {
      var count, iterator, date, index, yieldArgs;
      if (args === void 0) {
        args = {};
      }
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            count = args.take;
            delete args.take;
            iterator = this.occurrenceStream._run(args);
            date = iterator.next().value;
            index = 0;
            _a.label = 1;
          case 1:
            if (!(date && (count === undefined || count > index))) return [3 /*break*/, 3];
            date = date.add(this, 'generator');
            return [4 /*yield*/, this.normalizeRunOutput(date)];
          case 2:
            yieldArgs = _a.sent();
            date = iterator.next(yieldArgs).value;
            index++;
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/, undefined];
        }
      });
    };
    return ScheduleBase;
  })(OccurrenceGenerator);

  /**
   * An operator function which accepts a spread of occurrence generators
   * and removes their occurrences from the output.
   *
   * @param streams a spread of occurrence generators
   */
  function subtract() {
    var streams = [];
    for (var _i = 0; _i < arguments.length; _i++) {
      streams[_i] = arguments[_i];
    }
    return function(options) {
      return new SubtractOperator(streams, options);
    };
  }
  var SubtractOperator = /** @class */ (function(_super) {
    __extends(SubtractOperator, _super);
    function SubtractOperator() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    /** Not actually used but necessary for IRunnable interface */
    SubtractOperator.prototype.set = function(_, value) {
      return new SubtractOperator(
        this.streams.map(function(stream) {
          return stream.set('timezone', value);
        }),
        __assign(__assign({}, this.config), {
          base: this.config.base && this.config.base.set('timezone', value),
          timezone: value,
        }),
      );
    };
    SubtractOperator.prototype._run = function(args) {
      var inclusion, exclusion, yieldArgs;
      if (args === void 0) {
        args = {};
      }
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            if (!this.config.base) return [2 /*return*/];
            inclusion = new IterableWrapper(this.config.base, args);
            exclusion = new IterableWrapper(
              new AddOperator(this.streams, {
                timezone: this.config.timezone,
              }),
              args,
            );
            cycleStreams(inclusion, exclusion, args);
            _a.label = 1;
          case 1:
            if (!!inclusion.done) return [3 /*break*/, 3];
            return [4 /*yield*/, this.normalizeRunOutput(inclusion.value)];
          case 2:
            yieldArgs = _a.sent();
            if (!(yieldArgs && yieldArgs.skipToDate)) {
              inclusion.next();
            }
            cycleStreams(inclusion, exclusion, args, yieldArgs);
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/];
        }
      });
    };
    SubtractOperator.prototype.calculateIsInfinite = function() {
      return !!(this.config.base && this.config.base.isInfinite);
    };
    SubtractOperator.prototype.calculateHasDuration = function() {
      return !!(this.config.base && this.config.base.hasDuration);
    };
    return SubtractOperator;
  })(Operator);
  function cycleStreams(inclusion, exclusion, options, yieldArgs) {
    if (options === void 0) {
      options = {};
    }
    if (yieldArgs === void 0) {
      yieldArgs = {};
    }
    processYieldArgs([inclusion, exclusion], options, yieldArgs);
    iterateExclusion(inclusion, exclusion, options);
    while (!inclusion.done && !exclusion.done && inclusion.value.isEqual(exclusion.value)) {
      inclusion.next();
      iterateExclusion(inclusion, exclusion, options);
    }
  }
  function iterateExclusion(inclusion, exclusion, options) {
    if (options === void 0) {
      options = {};
    }
    if (options.reverse) {
      while (!exclusion.done && !inclusion.done && exclusion.value.isAfter(inclusion.value)) {
        exclusion.next();
      }
      return;
    }
    while (!exclusion.done && !inclusion.done && exclusion.value.isBefore(inclusion.value)) {
      exclusion.next();
    }
  }

  /**
   * An operator function which deduplicates an occurrence stream. Occurrence
   * `duration` is currently ignored.
   */
  function unique() {
    return function(options) {
      return new UniqueOperator([], options);
    };
  }
  var UniqueOperator = /** @class */ (function(_super) {
    __extends(UniqueOperator, _super);
    function UniqueOperator() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    /** Not actually used but necessary for IRunnable interface */
    UniqueOperator.prototype.set = function(_, value) {
      return new UniqueOperator(
        [],
        __assign(__assign({}, this.config), {
          base: this.config.base && this.config.base.set('timezone', value),
          timezone: value,
        }),
      );
    };
    UniqueOperator.prototype._run = function(args) {
      var stream, yieldArgs, lastValue;
      if (args === void 0) {
        args = {};
      }
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            if (!this.config.base) return [2 /*return*/];
            stream = new IterableWrapper(this.config.base, args);
            _a.label = 1;
          case 1:
            if (!!stream.done) return [3 /*break*/, 3];
            return [4 /*yield*/, this.normalizeRunOutput(stream.value)];
          case 2:
            yieldArgs = _a.sent();
            lastValue = stream.value;
            // iterate the current stream
            stream.next(yieldArgs);
            while (
              !(yieldArgs && yieldArgs.skipToDate) &&
              !stream.done &&
              stream.value.isEqual(lastValue)
            ) {
              stream.next();
            }
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/];
        }
      });
    };
    UniqueOperator.prototype.calculateIsInfinite = function() {
      return !!(this.config.base && this.config.base.isInfinite);
    };
    UniqueOperator.prototype.calculateHasDuration = function() {
      return !!(this.config.base && this.config.base.hasDuration);
    };
    return UniqueOperator;
  })(Operator);

  var Schedule = /** @class */ (function(_super) {
    __extends(Schedule, _super);
    /**
     * Create a new Schedule object with the specified options.
     *
     * The order of precidence for rrules, rdates, exrules, and exdates is:
     *
     * 1. rrules are included
     * 2. exrules are excluded
     * 3. rdates are included
     * 4. exdates are excluded
     *
     * ### Options
     *
     * - **timezone**: The timezone that yielded occurrences should be *displayed* in.
     *   Note, this one affects the *displayed* timezone of yielded occurrences.
     *   For rules, occurrences are first found using the unmodified rule
     *   config (including whatever timezone the `start` datetime is defined
     *   in), and then converted to the timezone specified here before being
     *   yielded. By default, the timezone is *local* time (`null`). So if you don't
     *   want your rules to be displayed in local time, you must supply a
     *   timezone argument.
     * - **data**: arbitrary data you can associate with this Schedule. This
     *   is the only mutable property of `Schedule` objects.
     * - **maxDuration**: currently unused.
     * - **rrules**: rules specifying when occurrences happen. See the "Rule Config"
     *   section below.
     * - **rdates**: individual dates that should be _included_ in the schedule.
     * - **exdates**: individual dates that should be _excluded_ from the schedule.
     * - **exrules**: rules specifying when occurrences shouldn't happen. See the
     *   "Rule Config" section below.
     *
     * ### Rule Config
     *
     * - #### frequency
     *
     *   The frequency rule part identifies the type of recurrence rule. Valid values
     *   include `"SECONDLY"`, `"MINUTELY"`, `"HOURLY"`, `"DAILY"`, `"WEEKLY"`,
     *   `"MONTHLY"`, or `"YEARLY"`.
     *
     * - #### start
     *
     *   The start of the rule (not necessarily the first occurrence).
     *   Either a `DateAdapter` instance, date object, or `DateTime` object.
     *   The type of date object depends on the `DateAdapter` class used for this
     *   `Rule`.
     *
     * - #### end?
     *
     *   The end of the rule (not necessarily the last occurrence).
     *   Either a `DateAdapter` instance, date object, or `DateTime` object.
     *   The type of date object depends on the `DateAdapter` class used for this
     *   `Rule`.
     *
     * - #### duration?
     *
     *   A length of time expressed in milliseconds.
     *
     * - #### interval?
     *
     *   The interval rule part contains a positive integer representing at
     *   which intervals the recurrence rule repeats. The default value is
     *   `1`, meaning every second for a SECONDLY rule, every minute for a
     *   MINUTELY rule, every hour for an HOURLY rule, every day for a
     *   DAILY rule, every week for a WEEKLY rule, every month for a
     *   MONTHLY rule, and every year for a YEARLY rule. For example,
     *   within a DAILY rule, a value of `8` means every eight days.
     *
     * - #### count?
     *
     *   The count rule part defines the number of occurrences at which to
     *   range-bound the recurrence. `count` and `end` are both two different
     *   ways of specifying how a recurrence completes.
     *
     * - #### weekStart?
     *
     *   The weekStart rule part specifies the day on which the workweek starts.
     *   Valid values are `"MO"`, `"TU"`, `"WE"`, `"TH"`, `"FR"`, `"SA"`, and `"SU"`.
     *   This is significant when a WEEKLY rule has an interval greater than 1,
     *   and a `byDayOfWeek` rule part is specified. The
     *   default value is `"MO"`.
     *
     * - #### bySecondOfMinute?
     *
     *   The bySecondOfMinute rule part expects an array of seconds
     *   within a minute. Valid values are 0 to 60.
     *
     * - #### byMinuteOfHour?
     *
     *   The byMinuteOfHour rule part expects an array of minutes within an hour.
     *   Valid values are 0 to 59.
     *
     * - #### byHourOfDay?
     *
     *   The byHourOfDay rule part expects an array of hours of the day.
     *   Valid values are 0 to 23.
     *
     * - #### byDayOfWeek?
     *
     *   *note: the byDayOfWeek rule part is kinda complex. Blame the ICAL spec.*
     *
     *   The byDayOfWeek rule part expects an array. Each array entry can
     *   be a day of the week (`"SU"`, `"MO"` , `"TU"`, `"WE"`, `"TH"`,
     *   `"FR"`, `"SA"`). If the rule's `frequency` is either MONTHLY or YEARLY,
     *   Any entry can also be a tuple where the first value of the tuple is a
     *   day of the week and the second value is an positive/negative integer
     *   (e.g. `["SU", 1]`). In this case, the number indicates the nth occurrence of
     *   the specified day within the MONTHLY or YEARLY rule.
     *
     *   The behavior of byDayOfWeek changes depending on the `frequency`
     *   of the rule.
     *
     *   Within a MONTHLY rule, `["MO", 1]` represents the first Monday
     *   within the month, whereas `["MO", -1]` represents the last Monday
     *   of the month.
     *
     *   Within a YEARLY rule, the numeric value in a byDayOfWeek tuple entry
     *   corresponds to an offset within the month when the byMonthOfYear rule part is
     *   present, and corresponds to an offset within the year otherwise.
     *
     *   Regardless of rule `frequency`, if a byDayOfWeek entry is a string
     *   (rather than a tuple), it means "all of these days" within the specified
     *   frequency (e.g. within a MONTHLY rule, `"MO"` represents all Mondays within
     *   the month).
     *
     * - #### byDayOfMonth?
     *
     *   The byDayOfMonth rule part expects an array of days
     *   of the month. Valid values are 1 to 31 or -31 to -1.
     *
     *   For example, -10 represents the tenth to the last day of the month.
     *   The byDayOfMonth rule part *must not* be specified when the rule's
     *   `frequency` is set to WEEKLY.
     *
     * - #### byMonthOfYear?
     *
     *   The byMonthOfYear rule part expects an array of months
     *   of the year. Valid values are 1 to 12.
     *
     */
    function Schedule(options) {
      var e_1, _a, e_2, _b;
      if (options === void 0) {
        options = {};
      }
      var _this = _super.call(this, options) || this;
      _this.rrules = [];
      _this.exrules = [];
      try {
        for (var _c = __values(['rrules', 'exrules']), _d = _c.next(); !_d.done; _d = _c.next()) {
          var prop = _d.value;
          var arg = options[prop];
          if (arg) {
            _this[prop] = arg.map(function(ruleArgs) {
              if (ruleArgs instanceof Rule) {
                return ruleArgs.set('timezone', _this.timezone);
              } else {
                return new Rule(ruleArgs, {
                  timezone: _this.timezone,
                });
              }
            });
          }
        }
      } catch (e_1_1) {
        e_1 = { error: e_1_1 };
      } finally {
        try {
          if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
        } finally {
          if (e_1) throw e_1.error;
        }
      }
      try {
        for (var _e = __values(['rdates', 'exdates']), _f = _e.next(); !_f.done; _f = _e.next()) {
          var prop = _f.value;
          var arg = options[prop];
          if (arg) {
            _this[prop] =
              arg instanceof Dates
                ? arg.set('timezone', _this.timezone)
                : new Dates({
                    dates: arg,
                    timezone: _this.timezone,
                  });
          } else {
            _this[prop] = new Dates({
              timezone: _this.timezone,
            });
          }
        }
      } catch (e_2_1) {
        e_2 = { error: e_2_1 };
      } finally {
        try {
          if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
        } finally {
          if (e_2) throw e_2.error;
        }
      }
      _this.hasDuration =
        _this.rrules.every(function(rule) {
          return rule.hasDuration;
        }) &&
        _this.exrules.every(function(rule) {
          return rule.hasDuration;
        }) &&
        _this.rdates.hasDuration &&
        _this.exdates.hasDuration;
      _this.isInfinite = _this.rrules.some(function(rule) {
        return rule.isInfinite;
      });
      _this.occurrenceStream = [
        add.apply(void 0, __spread(_this.rrules)),
        subtract.apply(void 0, __spread(_this.exrules)),
        add(_this.rdates),
        subtract(_this.exdates),
        unique(),
      ].reduce(function(prev, curr) {
        return curr({ base: prev, timezone: _this.timezone });
      }, undefined);
      return _this;
    }
    Schedule.prototype.add = function(prop, value) {
      var rrules = this.rrules.slice();
      var exrules = this.exrules.slice();
      var rdates = this.rdates;
      var exdates = this.exdates;
      switch (prop) {
        case 'rrule':
          rrules.push(value);
          break;
        case 'exrule':
          exrules.push(value);
          break;
        case 'rdate':
          rdates = this.rdates.add(value);
          break;
        case 'exdate':
          exdates = this.exdates.add(value);
          break;
      }
      return new Schedule({
        timezone: this.timezone,
        data: this.data,
        rrules: rrules,
        exrules: exrules,
        rdates: rdates,
        exdates: exdates,
      });
    };
    Schedule.prototype.remove = function(prop, value) {
      var rrules = this.rrules;
      var exrules = this.exrules;
      var rdates = this.rdates;
      var exdates = this.exdates;
      switch (prop) {
        case 'rrule':
          rrules = rrules.filter(function(rule) {
            return rule !== value;
          });
          break;
        case 'exrule':
          exrules = exrules.filter(function(rule) {
            return rule !== value;
          });
          break;
        case 'rdate':
          rdates = this.rdates.remove(value);
          break;
        case 'exdate':
          exdates = this.exdates.remove(value);
          break;
      }
      return new Schedule({
        timezone: this.timezone,
        data: this.data,
        rrules: rrules,
        exrules: exrules,
        rdates: rdates,
        exdates: exdates,
      });
    };
    Schedule.prototype.set = function(prop, value, options) {
      if (options === void 0) {
        options = {};
      }
      var timezone = this.timezone;
      var rrules = this.rrules;
      var exrules = this.exrules;
      var rdates = this.rdates;
      var exdates = this.exdates;
      switch (prop) {
        case 'timezone':
          if (value === this.timezone && !options.keepLocalTime) return this;
          else if (options.keepLocalTime) {
            rrules = rrules.map(function(rule) {
              return rule.set('timezone', value, options);
            });
            exrules = exrules.map(function(rule) {
              return rule.set('timezone', value, options);
            });
            rdates = rdates.set('timezone', value, options);
            exdates = exdates.set('timezone', value, options);
          }
          timezone = value;
          break;
        case 'rrules':
          rrules = value;
          break;
        case 'exrules':
          exrules = value;
          break;
        case 'rdates':
          rdates = value;
          break;
        case 'exdates':
          exdates = value;
          break;
      }
      return new Schedule({
        timezone: timezone,
        data: this.data,
        rrules: rrules,
        exrules: exrules,
        rdates: rdates,
        exdates: exdates,
      });
    };
    return Schedule;
  })(ScheduleBase);

  /**
   * An operator function, which takes a spread of occurrence generators and only
   * returns the dates which intersect every occurrence generator.
   *
   * Because it's possible for all the generators to never intersect,
   * and because the intersection operator can't detect this lack of intersection,
   * you must call `intersection()` with a `{maxFailedIterations: number}` argument.
   * For convenience, you can globally set `RScheduleConfig.defaultMaxFailedIterations`.
   * Without further information, I'd probably set `defaultMaxFailedIterations = 50`.
   *
   * The `maxFailedIterations` argument caps the number of iterations the operator will
   * run through without finding a single valid occurrence. If this number is reached, the operator will
   * stop iterating (preventing a possible infinite loop).
   *
   * - Note: `maxFailedIterations` caps the number of iterations which
   *   *fail to turn up a single valid occurrence*. Every time a valid occurrence is returned,
   *   the current iteration count is reset to 0.
   *
   */
  function intersection(args) {
    return function(options) {
      return new IntersectionOperator(args, options);
    };
  }
  var IntersectionOperator = /** @class */ (function(_super) {
    __extends(IntersectionOperator, _super);
    function IntersectionOperator(args, config) {
      var _this = _super.call(this, args.streams, config) || this;
      if (_this.isInfinite) {
        _this.maxFailedIterations =
          args.maxFailedIterations || IntersectionOperator.defaultMaxFailedIterations;
        if (!_this.maxFailedIterations) {
          throw new core.ArgumentError(
            'The IntersectionOperator must be provided ' +
              'a `maxFailedIterations` argument when it is built from schedules of infinite length. ' +
              'This argument is used to ensure that the IntersectionOperator does not enter ' +
              'an infinite loop because the underlying schedules never intersect. ' +
              'If the `maxFailedIterations` count is reached it will be assumed that ' +
              'all valid occurrences have been found and iteration will end without error.' +
              'Without additional information, "50" is probably a good ' +
              '`maxFailedIterations` value. ' +
              'If the schedules are not of infinite length, `maxFailedIterations` is ignored. ' +
              'Note also that you can provide a `defaultMaxFailedIterations` number via ' +
              'IntersectionOperator.defaultMaxFailedIterations.',
          );
        }
      }
      return _this;
    }
    IntersectionOperator.prototype.set = function(_, value) {
      return new IntersectionOperator(
        {
          maxFailedIterations: this.maxFailedIterations,
          streams: this.streams.map(function(stream) {
            return stream.set('timezone', value);
          }),
        },
        __assign(__assign({}, this.config), {
          base: this.config.base && this.config.base.set('timezone', value),
          timezone: value,
        }),
      );
    };
    IntersectionOperator.prototype._run = function(args) {
      var streams, hasEndDate, stream, yieldArgs, lastValidDate;
      if (args === void 0) {
        args = {};
      }
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            streams = this.streams.map(function(stream) {
              return new IterableWrapper(stream, args);
            });
            if (this.config.base) {
              streams.push(new IterableWrapper(this.config.base, args));
            }
            if (streams.length === 0) return [2 /*return*/];
            hasEndDate = !!(!this.isInfinite || args.reverse || args.end);
            if (
              !cycleStreams$1(
                streams,
                undefined,
                __assign(__assign({}, args), {
                  hasEndDate: hasEndDate,
                  iteration: 0,
                  maxIterations: this.maxFailedIterations,
                }),
              )
            ) {
              return [2 /*return*/];
            }
            stream = selectNextIterable(streams, args);
            _a.label = 1;
          case 1:
            if (!stream) return [3 /*break*/, 3];
            return [4 /*yield*/, this.normalizeRunOutput(stream.value)];
          case 2:
            yieldArgs = _a.sent();
            lastValidDate = stream.value;
            if (!(yieldArgs === null || yieldArgs === void 0 ? void 0 : yieldArgs.skipToDate)) {
              // iterate the current stream
              stream.next();
            }
            if (
              !cycleStreams$1(
                streams,
                lastValidDate,
                __assign(__assign({}, args), {
                  hasEndDate: hasEndDate,
                  iteration: 0,
                  maxIterations: this.maxFailedIterations,
                }),
                yieldArgs,
              )
            ) {
              return [2 /*return*/];
            }
            // The call to `cycleStreams()`, above, has already called
            // selectNextIterable with the `yieldArgs`
            stream = selectNextIterable(streams, args);
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/];
        }
      });
    };
    IntersectionOperator.prototype.calculateIsInfinite = function() {
      // Note: Array#every() === true when length === 0
      if (!this.config.base) {
        if (this.streams.length === 0) return false;
        return this.streams.every(function(stream) {
          return stream.isInfinite;
        });
      } else if (this.streams.length === 0) return this.config.base.isInfinite;
      return (
        this.config.base.isInfinite &&
        this.streams.every(function(stream) {
          return stream.isInfinite;
        })
      );
    };
    IntersectionOperator.prototype.calculateHasDuration = function() {
      var streamsDuration = this.streams.every(function(stream) {
        return stream.hasDuration;
      });
      if (!this.config.base) return streamsDuration;
      return this.config.base.hasDuration && streamsDuration;
    };
    return IntersectionOperator;
  })(Operator);
  function cycleStreams$1(streams, lastValidDate, options, yieldArgs) {
    var next = selectNextIterable(streams, options, yieldArgs);
    if (!next) return false;
    if (lastValidDate && next.value.isEqual(lastValidDate)) return true;
    if (
      streams.some(function(stream) {
        return stream.done;
      })
    )
      return false;
    if (
      streams.every(function(stream) {
        return stream.value.isEqual(next.value);
      })
    )
      return true;
    options.iteration++;
    if (options.maxIterations && !options.hasEndDate && options.iteration > options.maxIterations) {
      return false;
    }
    // Since not all of the streams are equal, we grab the last stream...
    var last = selectLastIterable(streams, options);
    // ...and skip all the other streams so they are equal or past the last one
    streams.forEach(function(stream) {
      // Because streams can have multiple, identical dates in a row,
      // we don't want to `skipToDate` if the provided date is equal to the current date.
      if (last.value.isEqual(stream.value)) return;
      stream.next({ skipToDate: last.value });
    });
    // then we repeat
    return cycleStreams$1(streams, lastValidDate, options);
  }

  var DurationIterableWrapper = /** @class */ (function(_super) {
    __extends(DurationIterableWrapper, _super);
    function DurationIterableWrapper(generator, runArgs) {
      var _this = _super.call(this, generator, runArgs) || this;
      _this.workingValue = _this.value;
      _this.next();
      return _this;
    }
    return DurationIterableWrapper;
  })(IterableWrapper);
  var MergeDurationOperatorError = /** @class */ (function(_super) {
    __extends(MergeDurationOperatorError, _super);
    function MergeDurationOperatorError() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    return MergeDurationOperatorError;
  })(Error);
  /**
   * An operator function which takes an occurrence stream with
   * `hasDuration === true` and merges occurrences which have overlapping
   * start and end times.
   *
   * Because it's possible for all the occurrences in the stream to have
   * overlapping start and end times, you must provide a `maxDuration`
   * argument that represents the maximum possible duration for a single
   * occurrence. If this duration is exceeded, a `MergeDurationOperatorError`
   * will be thrown.
   *
   * - For your convenience, you can globally set a default
   *   `MergeDurationOperator#maxDuration` via
   *   `RScheduleConfig.MergeDurationOperator.defaultMaxDuration`.
   *
   * Usage example:
   *
   * ```typescript
   * const MILLISECONDS_IN_HOUR = 1000 * 60 * 60;
   *
   * const dates = new Dates({
   *   dates: [
   *     new StandardDateAdapter(new Date(2010, 10, 10, 13), { duration: MILLISECONDS_IN_HOUR * 1 }),
   *     new StandardDateAdapter(new Date(2010, 10, 11, 13), { duration: MILLISECONDS_IN_HOUR * 2 }),
   *     new StandardDateAdapter(new Date(2010, 10, 11, 14), { duration: MILLISECONDS_IN_HOUR * 2 }),
   *     new StandardDateAdapter(new Date(2010, 10, 12, 13), { duration: MILLISECONDS_IN_HOUR * 1 }),
   *   ],
   *   dateAdpter: StandardDateAdapter,
   * }).pipe(
   *   mergeDuration({
   *     maxDuration: MILLISECONDS_IN_HOUR * 24
   *   })
   * )
   *
   * dates.occurrences().toArray() === [
   *   new StandardDateAdapter(new Date(2010, 10, 10, 13), { duration: MILLISECONDS_IN_HOUR * 1 }),
   *   new StandardDateAdapter(new Date(2010, 10, 11, 13), { duration: MILLISECONDS_IN_HOUR * 3 }),
   *   new StandardDateAdapter(new Date(2010, 10, 12, 13), { duration: MILLISECONDS_IN_HOUR * 1 }),
   * ]
   * ```
   */
  function mergeDuration(args) {
    return function(options) {
      return new MergeDurationOperator(args, options);
    };
  }
  var MergeDurationOperator = /** @class */ (function(_super) {
    __extends(MergeDurationOperator, _super);
    function MergeDurationOperator(args, config) {
      var _this = _super.call(this, [], config) || this;
      _this.maxDuration = args.maxDuration;
      if (config.base && !config.base.hasDuration) {
        throw new core.ArgumentError(
          'Base stream provided to MergeDurationOperator does not have an associated duration. ' +
            'The MergeDurationOperator can only be used with streams which have a duration. ',
        );
      }
      return _this;
    }
    /** Not actually used but necessary for IRunnable interface */
    MergeDurationOperator.prototype.set = function(_, value) {
      return new MergeDurationOperator(
        {
          maxDuration: this.maxDuration,
        },
        __assign(__assign({}, this.config), {
          base: this.config.base && this.config.base.set('timezone', value),
          timezone: value,
        }),
      );
    };
    MergeDurationOperator.prototype._run = function(args) {
      if (args === void 0) {
        args = {};
      }
      return args.reverse ? this.reverseRun(args) : this.forwardRun(args);
    };
    MergeDurationOperator.prototype.calculateIsInfinite = function() {
      return !!(this.config.base && this.config.base.isInfinite);
    };
    MergeDurationOperator.prototype.calculateHasDuration = function() {
      return true;
    };
    MergeDurationOperator.prototype.forwardRun = function(args) {
      var checkFromStart, checkFromEnd, stream, yieldArgs, diff;
      if (args === void 0) {
        args = {};
      }
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            if (!this.config.base) return [2 /*return*/];
            checkFromStart = args.start;
            if (args.start) {
              checkFromStart = args.start.subtract(this.maxDuration, 'millisecond');
            }
            checkFromEnd = args.end;
            if (args.end) {
              checkFromEnd = args.end.add(this.maxDuration, 'millisecond');
            }
            stream = new DurationIterableWrapper(
              this.config.base,
              __assign(__assign({}, args), { start: checkFromStart, end: checkFromEnd }),
            );
            _a.label = 1;
          case 1:
            if (!stream.workingValue) return [3 /*break*/, 3];
            // TODO(@john.carroll.p): figure out how to handle `DateTime#generators` for merged `DateTimes`
            while (!stream.done && stream.workingValue.end.isAfterOrEqual(stream.value)) {
              if (stream.workingValue.duration > this.maxDuration) {
                throw new MergeDurationOperatorError(
                  'MergeDurationOperatorError: Occurrence duration exceeded maxDuration of ' +
                    this.maxDuration,
                );
              }
              if (stream.value.end.isAfter(stream.workingValue.end)) {
                diff = stream.value.end.valueOf() - stream.workingValue.end.valueOf();
                stream.workingValue = stream.workingValue.set(
                  'duration',
                  stream.workingValue.duration + diff,
                );
              }
              stream.next();
            }
            // check to make sure the occurrence we are about to yield ends after the
            // provided start time.
            if (args.start && stream.workingValue.end.isBefore(args.start)) {
              stream.workingValue = stream.value;
              stream.next();
              return [3 /*break*/, 1];
            }
            // make sure the occurrence we are about to yield ends after the
            // provided skipToDate
            if (
              (yieldArgs === null || yieldArgs === void 0 ? void 0 : yieldArgs.skipToDate) &&
              stream.workingValue.end.isBefore(yieldArgs.skipToDate)
            ) {
              stream.workingValue = stream.value;
              stream.next();
              return [3 /*break*/, 1];
            }
            // make sure we are not after the user requested `end` time.
            if (args.end && stream.workingValue && stream.workingValue.isAfter(args.end)) {
              return [3 /*break*/, 3];
            }
            if (stream.workingValue.duration > this.maxDuration) {
              throw new MergeDurationOperatorError(
                'MergeDurationOperatorError: Occurrence duration exceeded maxDuration of ' +
                  this.maxDuration,
              );
            }
            return [4 /*yield*/, this.normalizeRunOutput(stream.workingValue)];
          case 2:
            yieldArgs = _a.sent();
            if (
              (yieldArgs === null || yieldArgs === void 0 ? void 0 : yieldArgs.skipToDate) &&
              stream.workingValue.isAfterOrEqual(yieldArgs.skipToDate)
            ) {
              throw new Error(
                'A provided `skipToDate` option must be greater than the last yielded date ' +
                  '(or smaller, in the case of reverse iteration)',
              );
            }
            stream.workingValue = stream.value;
            stream.next();
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/];
        }
      });
    };
    MergeDurationOperator.prototype.reverseRun = function(args) {
      var checkFromStart, checkFromEnd, stream, yieldArgs, diff;
      if (args === void 0) {
        args = {};
      }
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            if (!this.config.base) return [2 /*return*/];
            checkFromStart = args.start;
            if (args.start) {
              checkFromStart = args.start.subtract(this.maxDuration, 'millisecond');
            }
            checkFromEnd = args.end;
            if (args.end) {
              checkFromEnd = args.end.add(this.maxDuration, 'millisecond');
            }
            stream = new DurationIterableWrapper(
              this.config.base,
              __assign(__assign({}, args), { start: checkFromStart, end: checkFromEnd }),
            );
            _a.label = 1;
          case 1:
            if (!stream.workingValue) return [3 /*break*/, 3];
            // TODO(@john.carroll.p): figure out how to handle `DateTime#generators` for merged `DateTimes`
            while (!stream.done && stream.workingValue.isBeforeOrEqual(stream.value.end)) {
              if (stream.workingValue.duration > this.maxDuration) {
                throw new MergeDurationOperatorError(
                  'MergeDurationOperatorError: Occurrence duration exceeded maxDuration of ' +
                    this.maxDuration,
                );
              }
              if (
                stream.value.isBefore(stream.workingValue) ||
                stream.value.end.isAfter(stream.workingValue.end)
              ) {
                if (stream.value.end.isAfter(stream.workingValue.end)) {
                  // `stream.workingValue` is a subset of `stream.value`
                  // so simply replace `stream.workingValue` with `stream.value`
                  stream.workingValue = stream.value;
                } else {
                  diff = stream.workingValue.valueOf() - stream.value.valueOf();
                  stream.workingValue = stream.value.set(
                    'duration',
                    stream.workingValue.duration + diff,
                  );
                }
              }
              stream.next();
            }
            // check to make sure the occurrence we are about to yield starts before the
            // provided start time.
            if (args.start && stream.workingValue.end.isBefore(args.start)) {
              return [3 /*break*/, 3];
            }
            if (
              yieldArgs &&
              yieldArgs.skipToDate &&
              stream.workingValue.end.isBefore(yieldArgs.skipToDate)
            ) {
              stream.workingValue = stream.value;
              stream.next();
              return [3 /*break*/, 1];
            }
            // make sure we are not after the user requested `end` time.
            if (args.end && stream.workingValue && stream.workingValue.isAfter(args.end)) {
              stream.workingValue = stream.value;
              stream.next();
              return [3 /*break*/, 1];
            }
            if (stream.workingValue.duration > this.maxDuration) {
              throw new MergeDurationOperatorError(
                'MergeDurationOperatorError: Occurrence duration exceeded maxDuration of ' +
                  this.maxDuration,
              );
            }
            return [4 /*yield*/, this.normalizeRunOutput(stream.workingValue)];
          case 2:
            yieldArgs = _a.sent();
            if (
              yieldArgs &&
              yieldArgs.skipToDate &&
              stream.workingValue.end.isBeforeOrEqual(yieldArgs.skipToDate)
            ) {
              throw new Error(
                'A provided `skipToDate` option must be greater than the last yielded date ' +
                  '(or smaller, in the case of reverse iteration)',
              );
            }
            stream.workingValue = stream.value;
            stream.next();
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/];
        }
      });
    };
    return MergeDurationOperator;
  })(Operator);

  var SplitDurationOperatorError = /** @class */ (function(_super) {
    __extends(SplitDurationOperatorError, _super);
    function SplitDurationOperatorError() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    return SplitDurationOperatorError;
  })(Error);
  /**
   * An operator function which takes an occurrence stream with
   * `hasDuration === true` and passes occurrences through a splitting
   * function. One usecase for this operator is to dynamically break up
   * occurrences with a large duration into several smaller occurrences.
   *
   * You must provide a `maxDuration` argument that represents the
   * maximum possible duration for a single occurrence. If this
   * duration is exceeded, a `SplitDurationOperatorError` will be
   * thrown.
   *
   * - For your convenience, you can globally set a default
   *   `SplitDurationOperator#maxDuration` via
   *   `RScheduleConfig.SplitDurationOperator.defaultMaxDuration`.
   *
   * Usage example:
   *
   * ```typescript
   * const MILLISECONDS_IN_HOUR = 1000 * 60 * 60;
   *
   * const splitFn = (date: DateTime) => {
   *   if (date.duration > MILLISECONDS_IN_HOUR) {
   *     const diff = date.duration! / 2;
   *
   *     return [
   *       date.set('duration', diff),
   *       date.add(diff, 'millisecond').set('duration', diff),
   *     ];
   *   }
   *
   *   return [date];
   * };
   *
   * const dates = new Dates({
   *   dates: [
   *     new StandardDateAdapter(new Date(2010, 10, 10, 13), { duration: MILLISECONDS_IN_HOUR * 1 }),
   *     new StandardDateAdapter(new Date(2010, 10, 11, 13), { duration: MILLISECONDS_IN_HOUR * 2 }),
   *   ],
   *   dateAdpter: StandardDateAdapter,
   * }).pipe(
   *   splitDuration({
   *     splitFn,
   *     maxDuration: MILLISECONDS_IN_HOUR * 1
   *   })
   * )
   *
   * expect(dates.occurrences().toArray()).toEqual([
   *   new StandardDateAdapter(new Date(2010, 10, 10, 13), { duration: MILLISECONDS_IN_HOUR * 1 }),
   *   new StandardDateAdapter(new Date(2010, 10, 11, 13), { duration: MILLISECONDS_IN_HOUR * 1 }),
   *   new StandardDateAdapter(new Date(2010, 10, 11, 14), { duration: MILLISECONDS_IN_HOUR * 1 }),
   * ])
   * ```
   */
  function splitDuration(args) {
    return function(options) {
      return new SplitDurationOperator(args, options);
    };
  }
  var SplitDurationOperator = /** @class */ (function(_super) {
    __extends(SplitDurationOperator, _super);
    function SplitDurationOperator(args, config) {
      var _this = _super.call(this, [], config) || this;
      _this.splitFn = args.splitFn;
      _this.maxDuration = args.maxDuration;
      if (config.base && !config.base.hasDuration) {
        throw new core.ArgumentError(
          'Base stream provided to SplitDurationOperator does not have an associated duration. ' +
            'The SplitDurationOperator can only be used with streams which have a duration.',
        );
      }
      return _this;
    }
    /** Not actually used but necessary for IRunnable interface */
    SplitDurationOperator.prototype.set = function(_, value) {
      return new SplitDurationOperator(
        {
          maxDuration: this.maxDuration,
          splitFn: this.splitFn,
        },
        __assign(__assign({}, this.config), {
          base: this.config.base && this.config.base.set('timezone', value),
          timezone: value,
        }),
      );
    };
    SplitDurationOperator.prototype._run = function(args) {
      var reverse,
        checkFromStart,
        checkFromEnd,
        stream,
        yieldArgs,
        datesBucket,
        selectedDate,
        bucketIndex,
        selectedBucketIndex,
        dateIndex,
        selectedDateIndex,
        datesBucket_1,
        datesBucket_1_1,
        bucket,
        bucket_1,
        bucket_1_1,
        date,
        dateShouldComeNext;
      var e_1, _a, e_2, _b;
      if (args === void 0) {
        args = {};
      }
      return __generator(this, function(_c) {
        switch (_c.label) {
          case 0:
            if (!this.config.base) return [2 /*return*/];
            reverse = args.reverse || false;
            checkFromStart = args.start;
            if (args.start) {
              checkFromStart = args.start.subtract(this.maxDuration, 'millisecond');
            }
            checkFromEnd = args.end;
            if (args.end) {
              checkFromEnd = args.end.add(this.maxDuration, 'millisecond');
            }
            stream = new IterableWrapper(
              this.config.base,
              __assign(__assign({}, args), { start: checkFromStart, end: checkFromEnd }),
            );
            datesBucket = [];
            _c.label = 1;
          case 1:
            if (!(!stream.done || (datesBucket[0] && datesBucket[0][0]))) return [3 /*break*/, 3];
            /**
             * Example:
             * 10am - 2pm -> 10am - 12pm, 12pm - 2pm
             * 11am - 3pm -> 11am - 1pm, 1pm - 3pm
             * 2pm - 4pm -> 2pm - 3pm, 3pm - 4pm
             */
            if (!(datesBucket[0] && datesBucket[0][0])) {
              // we're out of dates
              datesBucket.push(this.splitDate(stream.value, reverse));
              stream.next();
            }
            while (
              !stream.done &&
              (reverse
                ? datesBucket[0].some(function(date) {
                    return date.isBeforeOrEqual(stream.value.end);
                  })
                : datesBucket[0].some(function(date) {
                    return date.isAfterOrEqual(stream.value);
                  }))
            ) {
              datesBucket.push(this.splitDate(stream.value, reverse));
              stream.next();
            }
            selectedDate = datesBucket[0] && datesBucket[0][0];
            bucketIndex = -1;
            selectedBucketIndex = 0;
            dateIndex = -1;
            selectedDateIndex = 0;
            try {
              // find the next date as well as its location in the datesBucket
              for (
                datesBucket_1 = ((e_1 = void 0), __values(datesBucket)),
                  datesBucket_1_1 = datesBucket_1.next();
                !datesBucket_1_1.done;
                datesBucket_1_1 = datesBucket_1.next()
              ) {
                bucket = datesBucket_1_1.value;
                bucketIndex++;
                dateIndex = -1;
                try {
                  for (
                    bucket_1 = ((e_2 = void 0), __values(bucket)), bucket_1_1 = bucket_1.next();
                    !bucket_1_1.done;
                    bucket_1_1 = bucket_1.next()
                  ) {
                    date = bucket_1_1.value;
                    dateIndex++;
                    dateShouldComeNext = void 0;
                    if (reverse) {
                      dateShouldComeNext =
                        date.isAfter(selectedDate) ||
                        (date.isEqual(selectedDate) && date.duration > selectedDate.duration);
                    } else {
                      dateShouldComeNext =
                        date.isBefore(selectedDate) ||
                        (date.isEqual(selectedDate) && date.duration < selectedDate.duration);
                    }
                    if (dateShouldComeNext) {
                      selectedDate = date;
                      selectedBucketIndex = bucketIndex;
                      selectedDateIndex = dateIndex;
                      break;
                    }
                  }
                } catch (e_2_1) {
                  e_2 = { error: e_2_1 };
                } finally {
                  try {
                    if (bucket_1_1 && !bucket_1_1.done && (_b = bucket_1.return)) _b.call(bucket_1);
                  } finally {
                    if (e_2) throw e_2.error;
                  }
                }
              }
            } catch (e_1_1) {
              e_1 = { error: e_1_1 };
            } finally {
              try {
                if (datesBucket_1_1 && !datesBucket_1_1.done && (_a = datesBucket_1.return))
                  _a.call(datesBucket_1);
              } finally {
                if (e_1) throw e_1.error;
              }
            }
            datesBucket[selectedBucketIndex].splice(selectedDateIndex, 1);
            if (datesBucket[selectedBucketIndex].length === 0) {
              datesBucket.splice(selectedBucketIndex, 1);
            }
            // If we've been yieldedArgs from the last cycle, check to see
            // that the selectedDate honors the `skipToDate` requirement
            // if not, discard this selectedDate
            if (
              yieldArgs &&
              yieldArgs.skipToDate &&
              selectedDate &&
              !datePastEnd(selectedDate, args) &&
              !datePastSkipToDate(selectedDate, yieldArgs.skipToDate, args)
            ) {
              return [3 /*break*/, 1];
            }
            // because we subtracted `maxDuration` to the base iterator's start time,
            // check to make sure the selectedDate we are about to yield should
            // actually be yielded (it may be before the provided `start` time).
            // If not, discard the selectedDate.
            if (args.start && selectedDate.end.isBefore(args.start)) {
              if (reverse) return [3 /*break*/, 3];
              return [3 /*break*/, 1];
            }
            // because we added `maxDuration` to the base iterator's end time,
            // check to make sure the selectedDate we are about to yield should
            // actually be yielded (it may be after the provided `end` time).
            // If not, end iteration.
            if (args.end && selectedDate.isAfter(args.end)) {
              if (reverse) return [3 /*break*/, 1];
              return [3 /*break*/, 3];
            }
            if (selectedDate.duration > this.maxDuration) {
              throw new SplitDurationOperatorError(
                'SplitDurationOperatorError: Occurrence duration exceeded maxDuration of ' +
                  this.maxDuration,
              );
            }
            return [4 /*yield*/, this.normalizeRunOutput(selectedDate)];
          case 2:
            yieldArgs = _c.sent();
            if (
              yieldArgs &&
              yieldArgs.skipToDate &&
              (args.reverse
                ? selectedDate.isBeforeOrEqual(yieldArgs.skipToDate)
                : selectedDate.isAfterOrEqual(yieldArgs.skipToDate))
            ) {
              throw new Error(
                'A provided `skipToDate` option must be greater than the last yielded date ' +
                  '(or smaller, in the case of reverse iteration)',
              );
            }
            return [3 /*break*/, 1];
          case 3:
            return [2 /*return*/];
        }
      });
    };
    SplitDurationOperator.prototype.calculateIsInfinite = function() {
      return !!(this.config.base && this.config.base.isInfinite);
    };
    SplitDurationOperator.prototype.calculateHasDuration = function() {
      return true;
    };
    SplitDurationOperator.prototype.splitDate = function(date, reverse) {
      var dates = this.splitFn(date);
      var valid;
      if (dates.length === 0) {
        valid = false;
      } else if (dates.length === 1) {
        valid = date.duration === dates[0].duration;
      } else {
        valid =
          date.duration ===
          dates.reduce(function(prev, curr) {
            return prev + curr.duration;
          }, 0);
      }
      if (!valid) {
        throw new Error(
          'The provided SplitDurationOperator split function ' +
            'must return an array of DateTimes with length > 0 ' +
            'where the total duration of the new dates equals the duration of ' +
            'the original date.',
        );
      }
      dates.sort(core.dateTimeSortComparer);
      if (reverse) {
        dates.reverse();
      }
      return dates;
    };
    return SplitDurationOperator;
  })(Operator);
  function datePastEnd(date, options) {
    return !!(options.reverse
      ? options.start && date.isBefore(options.start)
      : options.end && date.isAfter(options.end));
  }
  function datePastSkipToDate(date, skipToDate, options) {
    return !!(options.reverse ? skipToDate.isAfterOrEqual(date) : skipToDate.isBeforeOrEqual(date));
  }

  exports.AddOperator = AddOperator;
  exports.Calendar = Calendar;
  exports.Collection = Collection;
  exports.CollectionIterator = CollectionIterator;
  exports.Dates = Dates;
  exports.IntersectionOperator = IntersectionOperator;
  exports.MergeDurationOperator = MergeDurationOperator;
  exports.MergeDurationOperatorError = MergeDurationOperatorError;
  exports.OccurrenceGenerator = OccurrenceGenerator;
  exports.OccurrenceIterator = OccurrenceIterator;
  exports.Operator = Operator;
  exports.Rule = Rule;
  exports.RuleBase = RuleBase;
  exports.Schedule = Schedule;
  exports.ScheduleBase = ScheduleBase;
  exports.SplitDurationOperator = SplitDurationOperator;
  exports.SplitDurationOperatorError = SplitDurationOperatorError;
  exports.SubtractOperator = SubtractOperator;
  exports.UniqueOperator = UniqueOperator;
  exports.add = add;
  exports.intersection = intersection;
  exports.mergeDuration = mergeDuration;
  exports.splitDuration = splitDuration;
  exports.subtract = subtract;
  exports.unique = unique;

  Object.defineProperty(exports, '__esModule', { value: true });
});
