// @rschedule/core/umd/rules.js
(function(global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined'
    ? factory(exports, require('@rschedule/core'))
    : typeof define === 'function' && define.amd
    ? define(['exports', '@rschedule/core'], factory)
    : ((global = global || self), factory((global.rScheduleRules = {}), global.rSchedule));
})(this, function(exports, core) {
  'use strict';

  /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
  /* global Reflect, Promise */

  var extendStatics = function(d, b) {
    extendStatics =
      Object.setPrototypeOf ||
      ({ __proto__: [] } instanceof Array &&
        function(d, b) {
          d.__proto__ = b;
        }) ||
      function(d, b) {
        for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
      };
    return extendStatics(d, b);
  };

  function __extends(d, b) {
    extendStatics(d, b);
    function __() {
      this.constructor = d;
    }
    d.prototype = b === null ? Object.create(b) : ((__.prototype = b.prototype), new __());
  }

  function __values(o) {
    var m = typeof Symbol === 'function' && o[Symbol.iterator],
      i = 0;
    if (m) return m.call(o);
    return {
      next: function() {
        if (o && i >= o.length) o = void 0;
        return { value: o && o[i++], done: !o };
      },
    };
  }

  function __read(o, n) {
    var m = typeof Symbol === 'function' && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o),
      r,
      ar = [],
      e;
    try {
      while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    } catch (error) {
      e = { error: error };
    } finally {
      try {
        if (r && !r.done && (m = i['return'])) m.call(i);
      } finally {
        if (e) throw e.error;
      }
    }
    return ar;
  }

  function __spread() {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
  }

  var RecurrenceRule = /** @class */ (function() {
    function RecurrenceRule(processor) {
      this.processor = processor;
      this.start = processor.start;
      this.end = processor.end;
      this.options = processor.options;
    }
    return RecurrenceRule;
  })();

  var FrequencyRule = /** @class */ (function(_super) {
    __extends(FrequencyRule, _super);
    function FrequencyRule(processor, initDate) {
      var _this = _super.call(this, processor) || this;
      _this.initDate = initDate;
      _this.intervalUnit = core.freqToGranularity(_this.options.frequency);
      _this.firstIntervalStartDate = _this.normalizedStartDate(_this.options.start);
      _this.intervalStartDate = _this.firstIntervalStartDate;
      _this.intervalEndDate = _this.normalizedEndDate(_this.firstIntervalStartDate);
      _this.skipToInterval(_this.initDate);
      return _this;
    }
    FrequencyRule.prototype.run = function(date) {
      return this.validateDate(new core.ValidDateTime(date));
    };
    FrequencyRule.prototype.validateDate = function(arg) {
      var date = arg.date;
      if (arg instanceof core.ValidDateTime && this.dateIsWithinInterval(date)) {
        return arg;
      }
      this.skipToInterval(date);
      return new core.InvalidDateTime(
        // if the interval is 1, date will always be within the interval
        this.dateIsWithinInterval(date) ? date : this.setToCurrentInterval(),
      );
    };
    FrequencyRule.prototype.setToCurrentInterval = function() {
      return this.intervalStartDate;
    };
    FrequencyRule.prototype.normalizedStartDate = function(date) {
      if (this.options.frequency === 'WEEKLY') {
        return date.granularity('week', { weekStart: this.options.weekStart });
      }
      return date.granularity(this.intervalUnit);
    };
    FrequencyRule.prototype.normalizedEndDate = function(start) {
      switch (this.options.frequency) {
        case 'YEARLY':
          return start.add(1, 'year');
        case 'MONTHLY':
          return start.add(1, 'month');
        case 'WEEKLY':
          return start.add(1, 'week');
        case 'DAILY':
          return start.add(1, 'day');
        case 'HOURLY':
          return start.add(1, 'hour');
        case 'MINUTELY':
          return start.add(1, 'minute');
        case 'SECONDLY':
          return start.add(1, 'second');
        case 'MILLISECONDLY':
          return start.add(1, 'millisecond');
        default:
          throw new Error('Unknown frequency ' + this.options.frequency);
      }
    };
    FrequencyRule.prototype.skipToInterval = function(date) {
      var amount = this.intervalDifference(date);
      this.intervalStartDate = this.firstIntervalStartDate.add(amount, this.intervalUnit);
      this.intervalEndDate = this.normalizedEndDate(this.intervalStartDate);
    };
    FrequencyRule.prototype.dateIsWithinInterval = function(date) {
      return this.intervalStartDate.isBeforeOrEqual(date) && this.intervalEndDate.isAfter(date);
    };
    FrequencyRule.prototype.intervalDifference = function(date) {
      return intervalDifferenceBetweenDates({
        first: this.firstIntervalStartDate,
        second: date,
        unit: this.intervalUnit,
        interval: this.options.interval,
        weekStart: this.options.weekStart,
        direction: 'after',
      });
    };
    return FrequencyRule;
  })(RecurrenceRule);
  /**
   * Given the frequency (unit) and interval, this function finds
   * how many jumps forward the first date needs in order to equal
   * or exceed the second date.
   *
   * For example:
   *
   * 1. Unit is daily and interval is 1. The second date is 3 days
   *    after the first. This will return 3.
   * 2. Unit is yearly and interval is 1. The second date is 3 days
   *    after the first. This will return 0.
   * 3. Unit is yearly and interval is 3. The second date is 4 years
   *    after the first. This will return 6.
   */
  function intervalDifferenceBetweenDates(_a) {
    var first = _a.first,
      second = _a.second,
      unit = _a.unit,
      interval = _a.interval,
      weekStart = _a.weekStart,
      direction = _a.direction;
    var difference = (function() {
      var intervalDuration;
      var months;
      switch (unit) {
        case 'year':
          months = (second.get('year') - first.get('year')) * 12;
          months = months + second.get('month') - first.get('month');
          return Math.floor(months / 12);
        case 'month':
          months = (second.get('year') - first.get('year')) * 12;
          months = months + second.get('month') - first.get('month');
          return months;
        case 'week':
          first = first.granularity('week', { weekStart: weekStart });
          intervalDuration = core.DateAdapter.MILLISECONDS_IN_WEEK;
          break;
        case 'day':
          intervalDuration = core.DateAdapter.MILLISECONDS_IN_DAY;
          break;
        case 'hour':
          intervalDuration = core.DateAdapter.MILLISECONDS_IN_HOUR;
          break;
        case 'minute':
          intervalDuration = core.DateAdapter.MILLISECONDS_IN_MINUTE;
          break;
        case 'second':
          intervalDuration = core.DateAdapter.MILLISECONDS_IN_SECOND;
          break;
        case 'millisecond':
          intervalDuration = 1;
          break;
        default:
          throw new Error('Unexpected `unit` value');
      }
      var diff = second.valueOf() - first.valueOf();
      return Math.floor(diff / intervalDuration);
    })();
    var fn = direction === 'after' ? Math.ceil : Math.floor;
    difference = fn(difference / interval) * interval;
    return difference;
  }

  var RevFrequencyRule = /** @class */ (function(_super) {
    __extends(RevFrequencyRule, _super);
    function RevFrequencyRule() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    RevFrequencyRule.prototype.setToCurrentInterval = function() {
      return this.intervalEndDate.subtract(1, 'millisecond');
    };
    RevFrequencyRule.prototype.intervalDifference = function(date) {
      return intervalDifferenceBetweenDates({
        first: this.firstIntervalStartDate,
        second: date,
        unit: this.intervalUnit,
        interval: this.options.interval,
        weekStart: this.options.weekStart,
        direction: 'before',
      });
    };
    return RevFrequencyRule;
  })(FrequencyRule);

  var FREQUENCIES = [
    'MILLISECONDLY',
    'SECONDLY',
    'MINUTELY',
    'HOURLY',
    'DAILY',
    'WEEKLY',
    'MONTHLY',
    'YEARLY',
  ];
  var FrequencyRuleModule = {
    name: 'FrequencyRule',
    get: function(processor) {
      if (
        processor.options.byMillisecondOfSecond !== undefined ||
        processor.options.frequency !== 'MILLISECONDLY'
      ) {
        return null;
      }
      if (processor.reverse) return new RevFrequencyRule(processor, processor.end);
      return new FrequencyRule(processor, processor.start);
    },
    normalizeOptions: function(options, norm) {
      if (!FREQUENCIES.includes(options.frequency)) {
        throw new core.RuleOptionError('"frequency" must be one of ' + JSON.stringify(FREQUENCIES));
      }
      if (options.interval !== undefined) {
        if (!Number.isInteger(options.interval)) {
          throw new core.RuleOptionError('"interval" expects a whole number');
        }
        if (options.interval < 1) {
          throw new core.RuleOptionError('"interval" cannot be less than 1');
        }
      }
      if (options.weekStart !== undefined) {
        if (!core.DateAdapter.WEEKDAYS.includes(options.weekStart)) {
          throw new core.RuleOptionError(
            '"weekStart" must be one of ' + JSON.stringify(core.DateAdapter.WEEKDAYS),
          );
        }
      }
      norm.frequency = options.frequency;
      norm.interval = options.interval || 1;
      norm.weekStart = options.weekStart || 'MO';
    },
    deps: function() {
      return [FrequencyRuleModule];
    },
  };

  function ruleOptionFilled(option) {
    return Array.isArray(option) && option.length > 0;
  }

  var freqCache = new WeakMap();
  var RecurrenceRuleBase = /** @class */ (function(_super) {
    __extends(RecurrenceRuleBase, _super);
    function RecurrenceRuleBase(processor) {
      var _this = _super.call(this, processor) || this;
      if (!freqCache.has(_this.processor)) {
        freqCache.set(_this.processor, new FrequencyRule(processor, processor.start));
      }
      _this.frequency = freqCache.get(_this.processor);
      return _this;
    }
    RecurrenceRuleBase.prototype.validateDate = function(arg) {
      return this.frequency.validateDate(arg);
    };
    return RecurrenceRuleBase;
  })(RecurrenceRule);
  var RevRecurrenceRuleBase = /** @class */ (function(_super) {
    __extends(RevRecurrenceRuleBase, _super);
    function RevRecurrenceRuleBase(processor) {
      var _this = _super.call(this, processor) || this;
      if (!freqCache.has(_this.processor)) {
        freqCache.set(_this.processor, new RevFrequencyRule(processor, processor.end));
      }
      _this.frequency = freqCache.get(_this.processor);
      return _this;
    }
    RevRecurrenceRuleBase.prototype.validateDate = function(arg) {
      return this.frequency.validateDate(arg);
    };
    return RevRecurrenceRuleBase;
  })(RecurrenceRule);

  /**
   * Contains shared logic for ByHourOfDay, ByMinuteOfHour,
   * BySecondOfMinute, and ByMillisecondOfSecond reverse rule pipes
   */
  var RevByTimeOfUnitRule = /** @class */ (function(_super) {
    __extends(RevByTimeOfUnitRule, _super);
    function RevByTimeOfUnitRule() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    RevByTimeOfUnitRule.prototype.run = function(date) {
      var e_1, _a;
      // e.g. const currentTime = date.get('hour');
      var currentTime = date.get(this.granularity);
      try {
        // e.g. for (const time of this.options.byHourOfDay) {
        for (var _b = __values(this.option), _c = _b.next(); !_c.done; _c = _b.next()) {
          var time = _c.value;
          if (currentTime < time) continue;
          if (currentTime === time) {
            return this.validateDate(new core.ValidDateTime(date));
          }
          // e.g. return this.nextValidDate(args, date.endGranularity('day').set('hour', time));
          return this.validateDate(
            new core.InvalidDateTime(
              date.endGranularity(this.baseGranularity).set(this.granularity, time),
            ),
          );
        }
      } catch (e_1_1) {
        e_1 = { error: e_1_1 };
      } finally {
        try {
          if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        } finally {
          if (e_1) throw e_1.error;
        }
      }
      return this.validateDate(
        new core.InvalidDateTime(
          date
            // e.g. .endGranularity('day')
            .endGranularity(this.baseGranularity)
            // e.g. .subtract(1, 'day')
            .subtract(1, this.baseGranularity)
            // e.g. .set('hour', this.options.byHourOfDay[0]);
            .set(this.granularity, this.option[0]),
        ),
      );
    };
    return RevByTimeOfUnitRule;
  })(RevRecurrenceRuleBase);

  var RevByMillisecondOfSecondRule = /** @class */ (function(_super) {
    __extends(RevByMillisecondOfSecondRule, _super);
    function RevByMillisecondOfSecondRule() {
      var _this = (_super !== null && _super.apply(this, arguments)) || this;
      _this.baseGranularity = 'second';
      _this.granularity = 'millisecond';
      _this.option = _this.options.byMillisecondOfSecond.slice().reverse();
      return _this;
    }
    return RevByMillisecondOfSecondRule;
  })(RevByTimeOfUnitRule);

  /**
   * Contains shared logic for ByHourOfDay, ByMinuteOfHour,
   * BySecondOfMinute, and ByMillisecondOfSecond rule pipes
   */
  var ByTimeOfUnitRule = /** @class */ (function(_super) {
    __extends(ByTimeOfUnitRule, _super);
    function ByTimeOfUnitRule() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    ByTimeOfUnitRule.prototype.run = function(date) {
      var e_1, _a;
      // e.g. const currentTime = date.get('hour');
      var currentTime = date.get(this.granularity);
      try {
        // e.g. for (const time of this.options.byHourOfDay) {
        for (var _b = __values(this.option), _c = _b.next(); !_c.done; _c = _b.next()) {
          var time = _c.value;
          if (currentTime > time) continue;
          if (currentTime === time) {
            return this.validateDate(new core.ValidDateTime(date));
          }
          // e.g. return this.nextValidDate(args, date.granularity('day').set('hour', time));
          return this.validateDate(
            new core.InvalidDateTime(
              date.granularity(this.baseGranularity).set(this.granularity, time),
            ),
          );
        }
      } catch (e_1_1) {
        e_1 = { error: e_1_1 };
      } finally {
        try {
          if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        } finally {
          if (e_1) throw e_1.error;
        }
      }
      return this.validateDate(
        new core.InvalidDateTime(
          date
            // e.g. .granularity('day')
            .granularity(this.baseGranularity)
            // e.g. .add(1, 'day')
            .add(1, this.baseGranularity)
            // e.g. .set('hour', this.options.byHourOfDay[0]);
            .set(this.granularity, this.option[0]),
        ),
      );
    };
    return ByTimeOfUnitRule;
  })(RecurrenceRuleBase);

  var ByMillisecondOfSecondRule = /** @class */ (function(_super) {
    __extends(ByMillisecondOfSecondRule, _super);
    function ByMillisecondOfSecondRule() {
      var _this = (_super !== null && _super.apply(this, arguments)) || this;
      _this.baseGranularity = 'second';
      _this.granularity = 'millisecond';
      _this.option = _this.options.byMillisecondOfSecond;
      return _this;
    }
    return ByMillisecondOfSecondRule;
  })(ByTimeOfUnitRule);

  var ByMillisecondOfSecondRuleModule = {
    name: 'ByMillisecondOfSecond',
    get: function(processor) {
      if (processor.options.byMillisecondOfSecond === undefined) return null;
      if (processor.reverse) return new RevByMillisecondOfSecondRule(processor);
      return new ByMillisecondOfSecondRule(processor);
    },
    normalizeOptions: function(options, norm) {
      if (options.byMillisecondOfSecond !== undefined) {
        if (!ruleOptionFilled(options.byMillisecondOfSecond)) {
          throw new core.RuleOptionError('"byMillisecondOfSecond" expects a non-empty array');
        }
        if (
          options.byMillisecondOfSecond.some(function(num) {
            return num < 0 || num > 999;
          })
        ) {
          throw new core.RuleOptionError('"byMillisecondOfSecond" values must be >= 0 && <= 999');
        }
        norm.byMillisecondOfSecond = options.byMillisecondOfSecond;
        norm.byMillisecondOfSecond.sort(core.numberSortComparer);
      } else if (options.frequency !== 'MILLISECONDLY') {
        norm.byMillisecondOfSecond = [norm.start.get('millisecond')];
      }
    },
    deps: function() {
      return [FrequencyRuleModule, ByMillisecondOfSecondRuleModule];
    },
  };

  var RevBySecondOfMinuteRule = /** @class */ (function(_super) {
    __extends(RevBySecondOfMinuteRule, _super);
    function RevBySecondOfMinuteRule() {
      var _this = (_super !== null && _super.apply(this, arguments)) || this;
      _this.baseGranularity = 'minute';
      _this.granularity = 'second';
      _this.option = _this.options.bySecondOfMinute.slice().reverse();
      return _this;
    }
    return RevBySecondOfMinuteRule;
  })(RevByTimeOfUnitRule);

  var BySecondOfMinuteRule = /** @class */ (function(_super) {
    __extends(BySecondOfMinuteRule, _super);
    function BySecondOfMinuteRule() {
      var _this = (_super !== null && _super.apply(this, arguments)) || this;
      _this.baseGranularity = 'minute';
      _this.granularity = 'second';
      _this.option = _this.options.bySecondOfMinute;
      return _this;
    }
    return BySecondOfMinuteRule;
  })(ByTimeOfUnitRule);

  var BySecondOfMinuteRuleModule = {
    name: 'BySecondOfMinute',
    get: function(processor) {
      if (processor.options.bySecondOfMinute === undefined) return null;
      if (processor.reverse) return new RevBySecondOfMinuteRule(processor);
      return new BySecondOfMinuteRule(processor);
    },
    normalizeOptions: function(options, norm) {
      if (options.bySecondOfMinute !== undefined) {
        if (!ruleOptionFilled(options.bySecondOfMinute)) {
          throw new core.RuleOptionError('"bySecondOfMinute" expects a non-empty array');
        }
        if (
          options.bySecondOfMinute.some(function(num) {
            return num < 0 || num > 60;
          })
        ) {
          throw new core.RuleOptionError('"bySecondOfMinute" values must be >= 0 && <= 60');
        }
        norm.bySecondOfMinute = options.bySecondOfMinute;
        norm.bySecondOfMinute.sort(core.numberSortComparer);
      } else if (!['SECONDLY', 'MILLISECONDLY'].includes(options.frequency)) {
        norm.bySecondOfMinute = [norm.start.get('second')];
      }
    },
    deps: function() {
      return [FrequencyRuleModule, BySecondOfMinuteRuleModule, ByMillisecondOfSecondRuleModule];
    },
  };

  var RevByMinuteOfHourRule = /** @class */ (function(_super) {
    __extends(RevByMinuteOfHourRule, _super);
    function RevByMinuteOfHourRule() {
      var _this = (_super !== null && _super.apply(this, arguments)) || this;
      _this.baseGranularity = 'hour';
      _this.granularity = 'minute';
      _this.option = _this.options.byMinuteOfHour.slice().reverse();
      return _this;
    }
    return RevByMinuteOfHourRule;
  })(RevByTimeOfUnitRule);

  var ByMinuteOfHourRule = /** @class */ (function(_super) {
    __extends(ByMinuteOfHourRule, _super);
    function ByMinuteOfHourRule() {
      var _this = (_super !== null && _super.apply(this, arguments)) || this;
      _this.baseGranularity = 'hour';
      _this.granularity = 'minute';
      _this.option = _this.options.byMinuteOfHour;
      return _this;
    }
    return ByMinuteOfHourRule;
  })(ByTimeOfUnitRule);

  var ByMinuteOfHourRuleModule = {
    name: 'ByMinuteOfHour',
    get: function(processor) {
      if (processor.options.byMinuteOfHour === undefined) return null;
      if (processor.reverse) return new RevByMinuteOfHourRule(processor);
      return new ByMinuteOfHourRule(processor);
    },
    normalizeOptions: function(options, norm) {
      if (options.byMinuteOfHour !== undefined) {
        if (!ruleOptionFilled(options.byMinuteOfHour)) {
          throw new core.RuleOptionError('"byMinuteOfHour" expects a non-empty array');
        }
        if (
          options.byMinuteOfHour.some(function(num) {
            return num < 0 || num > 59;
          })
        ) {
          throw new core.RuleOptionError('"byMinuteOfHour" values must be >= 0 && <= 59');
        }
        norm.byMinuteOfHour = options.byMinuteOfHour;
        norm.byMinuteOfHour.sort(core.numberSortComparer);
      } else if (!['MINUTELY', 'SECONDLY', 'MILLISECONDLY'].includes(options.frequency)) {
        norm.byMinuteOfHour = [norm.start.get('minute')];
      }
    },
    deps: function() {
      return [
        FrequencyRuleModule,
        ByMinuteOfHourRuleModule,
        BySecondOfMinuteRuleModule,
        ByMillisecondOfSecondRuleModule,
      ];
    },
  };

  var RevByHourOfDayRule = /** @class */ (function(_super) {
    __extends(RevByHourOfDayRule, _super);
    function RevByHourOfDayRule() {
      var _this = (_super !== null && _super.apply(this, arguments)) || this;
      _this.baseGranularity = 'day';
      _this.granularity = 'hour';
      _this.option = _this.options.byHourOfDay.slice().reverse();
      return _this;
    }
    return RevByHourOfDayRule;
  })(RevByTimeOfUnitRule);

  var ByHourOfDayRule = /** @class */ (function(_super) {
    __extends(ByHourOfDayRule, _super);
    function ByHourOfDayRule() {
      var _this = (_super !== null && _super.apply(this, arguments)) || this;
      _this.baseGranularity = 'day';
      _this.granularity = 'hour';
      _this.option = _this.options.byHourOfDay;
      return _this;
    }
    return ByHourOfDayRule;
  })(ByTimeOfUnitRule);

  var ByHourOfDayRuleModule = {
    name: 'ByHourOfDay',
    get: function(processor) {
      if (processor.options.byHourOfDay === undefined) return null;
      if (processor.reverse) return new RevByHourOfDayRule(processor);
      return new ByHourOfDayRule(processor);
    },
    normalizeOptions: function(options, norm) {
      if (options.byHourOfDay !== undefined) {
        if (!ruleOptionFilled(options.byHourOfDay)) {
          throw new core.RuleOptionError('"byHourOfDay" expects a non-empty array');
        }
        if (
          options.byHourOfDay.some(function(num) {
            return num < 0 || num > 23;
          })
        ) {
          throw new core.RuleOptionError('"byHourOfDay" values must be >= 0 && <= 23');
        }
        norm.byHourOfDay = options.byHourOfDay;
        norm.byHourOfDay.sort(core.numberSortComparer);
      } else if (['YEARLY', 'MONTHLY', 'WEEKLY', 'DAILY'].includes(options.frequency)) {
        norm.byHourOfDay = [norm.start.get('hour')];
      }
    },
    deps: function() {
      return [
        FrequencyRuleModule,
        ByHourOfDayRuleModule,
        ByMinuteOfHourRuleModule,
        BySecondOfMinuteRuleModule,
        ByMillisecondOfSecondRuleModule,
      ];
    },
  };

  function getNextWeekday(date, weekday) {
    return date.add(core.getDifferenceBetweenWeekdays(date.get('weekday'), weekday), 'day');
  }

  function getNthWeekdayOfMonth(date, weekday, nth) {
    var base = date.set('day', 1);
    if (nth < 0) {
      base = base.add(1, 'month');
    }
    base = getNextWeekday(base, weekday);
    // when nth is negative, adding it will act as subtraction
    return nth < 0 ? base.add(nth, 'week') : base.add(nth - 1, 'week');
  }

  var ByDayOfMonthRule = /** @class */ (function(_super) {
    __extends(ByDayOfMonthRule, _super);
    function ByDayOfMonthRule() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    ByDayOfMonthRule.prototype.run = function(date) {
      var e_1, _a;
      var normalizedByDayOfMonth = normalizeByDayOfMonth(
        date,
        this.options.byDayOfMonth,
        this.options.byDayOfWeek,
      );
      var currentDay = date.get('day');
      try {
        for (
          var normalizedByDayOfMonth_1 = __values(normalizedByDayOfMonth),
            normalizedByDayOfMonth_1_1 = normalizedByDayOfMonth_1.next();
          !normalizedByDayOfMonth_1_1.done;
          normalizedByDayOfMonth_1_1 = normalizedByDayOfMonth_1.next()
        ) {
          var day = normalizedByDayOfMonth_1_1.value;
          if (currentDay > day) continue;
          if (currentDay === day) {
            return this.validateDate(new core.ValidDateTime(date));
          }
          return this.validateDate(
            new core.InvalidDateTime(date.granularity('month').set('day', day)),
          );
        }
      } catch (e_1_1) {
        e_1 = { error: e_1_1 };
      } finally {
        try {
          if (
            normalizedByDayOfMonth_1_1 &&
            !normalizedByDayOfMonth_1_1.done &&
            (_a = normalizedByDayOfMonth_1.return)
          )
            _a.call(normalizedByDayOfMonth_1);
        } finally {
          if (e_1) throw e_1.error;
        }
      }
      var next;
      var nextMonth = date;
      var index = 0;
      while (!next && index < 30) {
        nextMonth = nextMonth.granularity('month').add(1, 'month');
        next = normalizeByDayOfMonth(
          nextMonth,
          this.options.byDayOfMonth,
          this.options.byDayOfWeek,
        )[0];
        index++;
      }
      if (index >= 13) {
        throw new core.RecurrenceRuleError('byDayOfMonth Infinite while loop');
      }
      return this.validateDate(new core.InvalidDateTime(nextMonth.set('day', next)));
    };
    return ByDayOfMonthRule;
  })(RecurrenceRuleBase);
  /**
   * Does a few things:
   *
   * 1. filters out byDayOfMonth entries which are not applicable
   *    to current month
   * 2. negative entries to positive ones
   * 3. if a byDayOfWeek option is given, removes days which are
   *    not on the correct day of the week
   */
  function normalizeByDayOfMonth(date, byDayOfMonth, byDayOfWeek) {
    var lengthOfMonth = date.endGranularity('month').get('day');
    var normalizedByDayOfMonth = byDayOfMonth
      .filter(function(day) {
        return lengthOfMonth >= Math.abs(day);
      })
      .map(function(day) {
        return day > 0 ? day : lengthOfMonth + day + 1;
      });
    if (byDayOfWeek) {
      var base_1 = date.granularity('month');
      var filteredByDayOfMonth_1 = [];
      byDayOfWeek.forEach(function(entry) {
        if (typeof entry === 'string') {
          filteredByDayOfMonth_1.push.apply(
            filteredByDayOfMonth_1,
            __spread(
              normalizedByDayOfMonth.filter(function(day) {
                return base_1.set('day', day).get('weekday') === entry;
              }),
            ),
          );
          return;
        }
        var nthWeekdayOfMonth = getNthWeekdayOfMonth
          .apply(void 0, __spread([date], entry))
          .get('day');
        if (normalizedByDayOfMonth.includes(nthWeekdayOfMonth)) {
          filteredByDayOfMonth_1.push(nthWeekdayOfMonth);
        }
      });
      normalizedByDayOfMonth = Array.from(new Set(filteredByDayOfMonth_1));
    }
    return normalizedByDayOfMonth.sort(function(a, b) {
      if (a > b) return 1;
      if (a < b) return -1;
      else return 0;
    });
  }

  var RevByDayOfMonthRule = /** @class */ (function(_super) {
    __extends(RevByDayOfMonthRule, _super);
    function RevByDayOfMonthRule() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    RevByDayOfMonthRule.prototype.run = function(date) {
      var e_1, _a;
      var normalizedByDayOfMonth = normalizeByDayOfMonth(
        date,
        this.options.byDayOfMonth,
        this.options.byDayOfWeek,
      ).reverse();
      var currentDay = date.get('day');
      try {
        for (
          var normalizedByDayOfMonth_1 = __values(normalizedByDayOfMonth),
            normalizedByDayOfMonth_1_1 = normalizedByDayOfMonth_1.next();
          !normalizedByDayOfMonth_1_1.done;
          normalizedByDayOfMonth_1_1 = normalizedByDayOfMonth_1.next()
        ) {
          var day = normalizedByDayOfMonth_1_1.value;
          if (currentDay < day) continue;
          if (currentDay === day) {
            return this.validateDate(new core.ValidDateTime(date));
          }
          return this.validateDate(
            new core.InvalidDateTime(date.endGranularity('month').set('day', day)),
          );
        }
      } catch (e_1_1) {
        e_1 = { error: e_1_1 };
      } finally {
        try {
          if (
            normalizedByDayOfMonth_1_1 &&
            !normalizedByDayOfMonth_1_1.done &&
            (_a = normalizedByDayOfMonth_1.return)
          )
            _a.call(normalizedByDayOfMonth_1);
        } finally {
          if (e_1) throw e_1.error;
        }
      }
      var next;
      var nextMonth = date;
      var index = 0;
      while (!next && index < 30) {
        nextMonth = nextMonth.endGranularity('month').subtract(1, 'month');
        next = normalizeByDayOfMonth(
          nextMonth,
          this.options.byDayOfMonth,
          this.options.byDayOfWeek,
        ).pop();
        index++;
      }
      if (index >= 13) {
        throw new core.RecurrenceRuleError('byDayOfMonth Infinite while loop');
      }
      return this.validateDate(new core.InvalidDateTime(nextMonth.set('day', next)));
    };
    return RevByDayOfMonthRule;
  })(RevRecurrenceRuleBase);

  var ByDayOfMonthRuleModule = {
    name: 'ByDayOfMonth',
    get: function(processor) {
      if (processor.options.byDayOfMonth === undefined) return null;
      if (processor.reverse) return new RevByDayOfMonthRule(processor);
      return new ByDayOfMonthRule(processor);
    },
    normalizeOptions: function(options, norm) {
      if (options.byDayOfMonth !== undefined) {
        if (options.frequency === 'WEEKLY') {
          throw new core.RuleOptionError(
            'when "frequency" is "WEEKLY", "byDayOfMonth" cannot be present',
          );
        }
        if (!ruleOptionFilled(options.byDayOfMonth)) {
          throw new core.RuleOptionError('"byDayOfMonth" expects a non-empty array');
        }
        if (
          options.byDayOfMonth.some(function(num) {
            return num === 0 || num < -31 || num > 31;
          })
        ) {
          throw new core.RuleOptionError(
            '"byDayOfMonth" values must be `num !== 0 && num <= 31 && num >= -31`',
          );
        }
        norm.byDayOfMonth = options.byDayOfMonth.slice();
      } else if (
        !(ruleOptionFilled(options.byDayOfWeek) || ruleOptionFilled(options.byDayOfYear)) &&
        ['YEARLY', 'MONTHLY'].includes(options.frequency)
      ) {
        norm.byDayOfMonth = [norm.start.get('day')];
      }
    },
    deps: function() {
      return [
        FrequencyRuleModule,
        ByDayOfMonthRuleModule,
        ByHourOfDayRuleModule,
        ByMinuteOfHourRuleModule,
        BySecondOfMinuteRuleModule,
        ByMillisecondOfSecondRuleModule,
      ];
    },
  };

  function getNthWeekdayOfYear(date, weekday, nth) {
    var base = date.set('month', 1).set('day', 1);
    if (nth < 0) {
      base = base.add(1, 'year');
    }
    base = getNextWeekday(base, weekday);
    // when nth is negative, adding it will act as subtraction
    return nth < 0 ? base.add(nth, 'week') : base.add(nth - 1, 'week');
  }

  function getPreviousWeekday(date, weekday) {
    var diff = core.getDifferenceBetweenWeekdays(date.get('weekday'), weekday);
    return date.subtract(diff === 0 ? 0 : 7 - diff, 'day');
  }

  var RevByDayOfWeekRule = /** @class */ (function(_super) {
    __extends(RevByDayOfWeekRule, _super);
    function RevByDayOfWeekRule() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    RevByDayOfWeekRule.prototype.run = function(date) {
      if (this.options.frequency === 'YEARLY') {
        return this.options.byMonthOfYear === undefined
          ? this.expandYearly(date)
          : this.expandMonthly(date);
      } else if (this.options.frequency === 'MONTHLY') {
        return this.expandMonthly(date);
      }
      return this.expand(date);
    };
    RevByDayOfWeekRule.prototype.expandYearly = function(date) {
      var next = getPrevWeekdaysOfYear(date, this.options.byDayOfWeek)[0];
      var index = 0;
      var base = date;
      // If we can't find a valid date this year,
      // search the previous year. Only search the past 28 years.
      // (the calendar repeats on a 28 year cycle, according
      // to the internet).
      while (!next && index < 28) {
        index++;
        base = base.granularity('year').subtract(1, 'millisecond');
        next = getPrevWeekdaysOfYear(base, this.options.byDayOfWeek)[0];
      }
      if (!next) {
        throw new core.RecurrenceRuleError(
          'The byDayOfWeek rule appears to contain an impossible combination',
        );
      }
      return this.result(date, next);
    };
    RevByDayOfWeekRule.prototype.expandMonthly = function(date) {
      var next = getPrevWeekdaysOfMonth(date, this.options.byDayOfWeek)[0];
      var index = 0;
      var base = date;
      // TODO: performance improvment
      // If, in the first year, a match isn't found, we should be able to
      // jumpt to the previous leap year and check that. Or, if already on
      // a leap year, we can just error immediately.
      // If we can't find a valid date this month,
      // search the previous month. Only search the past 4 years
      // (to account for leap year).
      while (!next && index < 50) {
        index++;
        base = base.granularity('month').subtract(1, 'millisecond');
        next = getPrevWeekdaysOfMonth(base, this.options.byDayOfWeek)[0];
      }
      if (!next) {
        throw new core.RecurrenceRuleError(
          'The byDayOfWeek rule appears to contain an impossible combination',
        );
      }
      return this.result(date, next);
    };
    RevByDayOfWeekRule.prototype.expand = function(date) {
      var next = this.options.byDayOfWeek
        .map(function(weekday) {
          return getPreviousWeekday(date, weekday);
        })
        .sort(core.dateTimeSortComparer)
        .pop();
      return this.result(date, next);
    };
    RevByDayOfWeekRule.prototype.result = function(date, next) {
      if (next.isEqual(date)) {
        return this.validateDate(new core.ValidDateTime(date));
      }
      return this.validateDate(new core.InvalidDateTime(next.endGranularity('day')));
    };
    return RevByDayOfWeekRule;
  })(RevRecurrenceRuleBase);
  /** For each byDayOfWeek entry, find the previous DateTime */
  function getPrevWeekdaysOfYear(date, byDayOfWeek) {
    var normalizedNthWeekdaysOfYear = byDayOfWeek
      .filter(function(entry) {
        return Array.isArray(entry);
      })
      .map(function(entry) {
        return getNthWeekdayOfYear.apply(void 0, __spread([date], entry));
      });
    var normalizedNextWeekdays = byDayOfWeek
      .filter(function(entry) {
        return typeof entry === 'string';
      })
      .map(function(weekday) {
        return getPreviousWeekday(date, weekday);
      })
      .filter(function(entry) {
        return entry.get('year') === date.get('year');
      });
    return core
      .uniqDateTimes(__spread(normalizedNthWeekdaysOfYear, normalizedNextWeekdays))
      .filter(function(entry) {
        return entry.isBeforeOrEqual(date);
      })
      .sort(core.dateTimeSortComparer)
      .reverse();
  }
  /** For each byDayOfWeek entry, find the previous DateTime */
  function getPrevWeekdaysOfMonth(date, byDayOfWeek) {
    var normalizedNthWeekdaysOfMonth = byDayOfWeek
      .filter(function(entry) {
        return Array.isArray(entry);
      })
      .map(function(entry) {
        return getNthWeekdayOfMonth.apply(void 0, __spread([date], entry));
      });
    var normalizedNextWeekdays = byDayOfWeek
      .filter(function(entry) {
        return typeof entry === 'string';
      })
      .map(function(weekday) {
        return getPreviousWeekday(date, weekday);
      })
      .filter(function(entry) {
        return entry.get('year') === date.get('year') && entry.get('month') === date.get('month');
      });
    return core
      .uniqDateTimes(__spread(normalizedNthWeekdaysOfMonth, normalizedNextWeekdays))
      .filter(function(entry) {
        return entry.isBeforeOrEqual(date);
      })
      .sort(core.dateTimeSortComparer)
      .reverse();
  }

  var ByDayOfWeekRule = /** @class */ (function(_super) {
    __extends(ByDayOfWeekRule, _super);
    function ByDayOfWeekRule() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    ByDayOfWeekRule.prototype.run = function(date) {
      if (this.options.frequency === 'YEARLY') {
        return this.options.byMonthOfYear === undefined
          ? this.expandYearly(date)
          : this.expandMonthly(date);
      } else if (this.options.frequency === 'MONTHLY') {
        return this.expandMonthly(date);
      }
      return this.expand(date);
    };
    ByDayOfWeekRule.prototype.expandYearly = function(date) {
      var next = getNextWeekdaysOfYear(date, this.options.byDayOfWeek)[0];
      var index = 0;
      var base = date;
      // If we can't find a valid date this year,
      // search next year. Only search the next 28 years.
      // (the calendar repeats on a 28 year cycle, according
      // to the internet).
      while (!next && index < 28) {
        index++;
        base = base.granularity('year').add(1, 'year');
        next = getNextWeekdaysOfYear(base, this.options.byDayOfWeek)[0];
      }
      if (!next) {
        throw new core.RecurrenceRuleError(
          'The byDayOfWeek rule appears to contain an impossible combination',
        );
      }
      return this.result(date, next);
    };
    ByDayOfWeekRule.prototype.expandMonthly = function(date) {
      var next = getNextWeekdaysOfMonth(date, this.options.byDayOfWeek)[0];
      var index = 0;
      var base = date;
      // TODO: performance improvment
      // If, in the first year, a match isn't found, we should be able to
      // jumpt to the next leap year and check that. Or, if already on
      // a leap year, we can just error immediately.
      // If we can't find a valid date this month,
      // search the next month. Only search the next 4 years
      // (to account for leap year).
      while (!next && index < 50) {
        index++;
        base = base.granularity('month').add(1, 'month');
        next = getNextWeekdaysOfMonth(base, this.options.byDayOfWeek)[0];
      }
      if (!next) {
        throw new core.RecurrenceRuleError(
          'The byDayOfWeek rule appears to contain an impossible combination',
        );
      }
      return this.result(date, next);
    };
    ByDayOfWeekRule.prototype.expand = function(date) {
      var next = this.options.byDayOfWeek
        .map(function(weekday) {
          return getNextWeekday(date, weekday);
        })
        .sort(core.dateTimeSortComparer)[0];
      return this.result(date, next);
    };
    ByDayOfWeekRule.prototype.result = function(date, next) {
      if (next.isEqual(date)) {
        return this.validateDate(new core.ValidDateTime(date));
      }
      return this.validateDate(new core.InvalidDateTime(next.granularity('day')));
    };
    return ByDayOfWeekRule;
  })(RecurrenceRuleBase);
  /** For each byDayOfWeek entry, find the next DateTime */
  function getNextWeekdaysOfYear(date, byDayOfWeek) {
    var normalizedNthWeekdaysOfYear = byDayOfWeek
      .filter(function(entry) {
        return Array.isArray(entry);
      })
      .map(function(entry) {
        return getNthWeekdayOfYear.apply(void 0, __spread([date], entry));
      });
    var normalizedNextWeekdays = byDayOfWeek
      .filter(function(entry) {
        return typeof entry === 'string';
      })
      .map(function(weekday) {
        return getNextWeekday(date, weekday);
      })
      .filter(function(entry) {
        return entry.get('year') === date.get('year');
      });
    return core
      .uniqDateTimes(__spread(normalizedNthWeekdaysOfYear, normalizedNextWeekdays))
      .filter(function(entry) {
        return entry.isAfterOrEqual(date);
      })
      .sort(core.dateTimeSortComparer);
  }
  /** For each byDayOfWeek entry, find the next DateTime */
  function getNextWeekdaysOfMonth(date, byDayOfWeek) {
    var normalizedNthWeekdaysOfMonth = byDayOfWeek
      .filter(function(entry) {
        return Array.isArray(entry);
      })
      .map(function(entry) {
        return getNthWeekdayOfMonth.apply(void 0, __spread([date], entry));
      });
    var normalizedNextWeekdays = byDayOfWeek
      .filter(function(entry) {
        return typeof entry === 'string';
      })
      .map(function(weekday) {
        return getNextWeekday(date, weekday);
      })
      .filter(function(entry) {
        return entry.get('year') === date.get('year') && entry.get('month') === date.get('month');
      });
    return core
      .uniqDateTimes(__spread(normalizedNthWeekdaysOfMonth, normalizedNextWeekdays))
      .filter(function(entry) {
        return entry.isAfterOrEqual(date);
      })
      .sort(core.dateTimeSortComparer);
  }

  var ByDayOfWeekRuleModule = {
    name: 'ByDayOfWeek',
    get: function(processor) {
      if (processor.options.byDayOfWeek === undefined) return null;
      if (processor.reverse) return new RevByDayOfWeekRule(processor);
      return new ByDayOfWeekRule(processor);
    },
    normalizeOptions: function(options, norm) {
      if (options.byDayOfWeek !== undefined) {
        if (!ruleOptionFilled(options.byDayOfWeek)) {
          throw new core.RuleOptionError('"byDayOfWeek" expects a non-empty array');
        }
        var invalidWeeday = options.byDayOfWeek.find(function(day) {
          return Array.isArray(day)
            ? !core.DateAdapter.WEEKDAYS.includes(day[0])
            : !core.DateAdapter.WEEKDAYS.includes(day);
        });
        if (invalidWeeday) {
          throw new core.RuleOptionError(
            '"byDayOfWeek" expects weedays in the form ' +
              (JSON.stringify(core.DateAdapter.WEEKDAYS) +
                ' but "' +
                invalidWeeday +
                '" was provided'),
          );
        }
        if (
          !['YEARLY', 'MONTHLY'].includes(options.frequency) &&
          options.byDayOfWeek.some(function(weekday) {
            return Array.isArray(weekday);
          })
        ) {
          throw new core.RuleOptionError(
            '"byDayOfWeek" can only include a numeric value (i.e. `[string, number]`) when the "frequency" is ' +
              'either "MONTHLY" or "YEARLY"',
          );
        }
        if (
          options.frequency === 'MONTHLY' &&
          options.byDayOfWeek.some(function(weekday) {
            return (
              Array.isArray(weekday) && (weekday[1] < -31 || weekday[1] === 0 || weekday[1] > 31)
            );
          })
        ) {
          throw new core.RuleOptionError(
            'when "frequency" is "MONTHLY", each "byDayOfWeek" can optionally only' +
              ' have a numeric value >= -31 and <= 31 and !== 0',
          );
        }
        if (
          options.frequency === 'YEARLY' &&
          options.byDayOfWeek.some(function(weekday) {
            return (
              Array.isArray(weekday) && (weekday[1] < -366 || weekday[1] === 0 || weekday[1] > 366)
            );
          })
        ) {
          throw new core.RuleOptionError(
            'when "frequency" is "YEARLY", each "byDayOfWeek" can optionally only' +
              ' have a numeric value >= -366 and <= 366 and !== 0',
          );
        }
        norm.byDayOfWeek = options.byDayOfWeek;
      } else if (!ruleOptionFilled(options.byDayOfMonth) && options.frequency === 'WEEKLY') {
        norm.byDayOfWeek = [norm.start.get('weekday')];
      }
    },
    deps: function() {
      return [
        FrequencyRuleModule,
        ByDayOfWeekRuleModule,
        ByHourOfDayRuleModule,
        ByMinuteOfHourRuleModule,
        BySecondOfMinuteRuleModule,
        ByMillisecondOfSecondRuleModule,
      ];
    },
  };

  var RevByMonthOfYearRule = /** @class */ (function(_super) {
    __extends(RevByMonthOfYearRule, _super);
    function RevByMonthOfYearRule() {
      var _this = (_super !== null && _super.apply(this, arguments)) || this;
      _this.option = _this.processor.options.byMonthOfYear.slice().reverse();
      return _this;
    }
    RevByMonthOfYearRule.prototype.run = function(date) {
      var e_1, _a;
      var currentMonth = date.get('month');
      try {
        for (var _b = __values(this.option), _c = _b.next(); !_c.done; _c = _b.next()) {
          var month = _c.value;
          if (currentMonth < month) continue;
          if (currentMonth === month) {
            return this.validateDate(new core.ValidDateTime(date));
          }
          return this.validateDate(
            new core.InvalidDateTime(date.endGranularity('year').set('month', month)),
          );
        }
      } catch (e_1_1) {
        e_1 = { error: e_1_1 };
      } finally {
        try {
          if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        } finally {
          if (e_1) throw e_1.error;
        }
      }
      return this.validateDate(
        new core.InvalidDateTime(
          date
            .endGranularity('year')
            .subtract(1, 'year')
            .set('month', this.option[0]),
        ),
      );
    };
    return RevByMonthOfYearRule;
  })(RevRecurrenceRuleBase);

  var ByMonthOfYearRule = /** @class */ (function(_super) {
    __extends(ByMonthOfYearRule, _super);
    function ByMonthOfYearRule() {
      return (_super !== null && _super.apply(this, arguments)) || this;
    }
    ByMonthOfYearRule.prototype.run = function(date) {
      var e_1, _a;
      var currentMonth = date.get('month');
      try {
        for (
          var _b = __values(this.options.byMonthOfYear), _c = _b.next();
          !_c.done;
          _c = _b.next()
        ) {
          var month = _c.value;
          if (currentMonth > month) continue;
          if (currentMonth === month) {
            return this.validateDate(new core.ValidDateTime(date));
          }
          return this.validateDate(
            new core.InvalidDateTime(date.granularity('year').set('month', month)),
          );
        }
      } catch (e_1_1) {
        e_1 = { error: e_1_1 };
      } finally {
        try {
          if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        } finally {
          if (e_1) throw e_1.error;
        }
      }
      return this.validateDate(
        new core.InvalidDateTime(
          date
            .granularity('year')
            .add(1, 'year')
            .set('month', this.options.byMonthOfYear[0]),
        ),
      );
    };
    return ByMonthOfYearRule;
  })(RecurrenceRuleBase);

  var ByMonthOfYearRuleModule = {
    name: 'ByMonthOfYear',
    get: function(processor) {
      if (processor.options.byMonthOfYear === undefined) return null;
      if (processor.reverse) {
        return new RevByMonthOfYearRule(processor);
      }
      return new ByMonthOfYearRule(processor);
    },
    normalizeOptions: function(options, norm) {
      if (options.byMonthOfYear !== undefined) {
        if (!ruleOptionFilled(options.byMonthOfYear)) {
          throw new core.RuleOptionError('"byMonthOfYear" expects a non-empty array');
        }
        if (
          options.byMonthOfYear.some(function(num) {
            return num < 1 || num > 12;
          })
        ) {
          throw new core.RuleOptionError('"byMonthOfYear" values must be `num >= 1 && num >= 12`');
        }
        norm.byMonthOfYear = options.byMonthOfYear.slice();
        norm.byMonthOfYear.sort(core.numberSortComparer);
      } else if (
        !(
          ruleOptionFilled(options.byDayOfMonth) ||
          ruleOptionFilled(options.byDayOfWeek) ||
          ruleOptionFilled(options.byDayOfYear)
        ) &&
        options.frequency === 'YEARLY'
      ) {
        norm.byMonthOfYear = [norm.start.get('month')];
      }
    },
    deps: function() {
      return [
        FrequencyRuleModule,
        ByMonthOfYearRuleModule,
        ByDayOfMonthRuleModule,
        ByHourOfDayRuleModule,
        ByMinuteOfHourRuleModule,
        BySecondOfMinuteRuleModule,
        ByMillisecondOfSecondRuleModule,
      ];
    },
  };

  var ICAL_RULES = [
    FrequencyRuleModule,
    ByMonthOfYearRuleModule,
    ByDayOfMonthRuleModule,
    ByDayOfWeekRuleModule,
    ByHourOfDayRuleModule,
    ByMinuteOfHourRuleModule,
    BySecondOfMinuteRuleModule,
    ByMillisecondOfSecondRuleModule,
  ];

  exports.ByDayOfMonthRule = ByDayOfMonthRule;
  exports.ByDayOfMonthRuleModule = ByDayOfMonthRuleModule;
  exports.ByDayOfWeekRule = ByDayOfWeekRule;
  exports.ByDayOfWeekRuleModule = ByDayOfWeekRuleModule;
  exports.ByHourOfDayRule = ByHourOfDayRule;
  exports.ByHourOfDayRuleModule = ByHourOfDayRuleModule;
  exports.ByMillisecondOfSecondRule = ByMillisecondOfSecondRule;
  exports.ByMillisecondOfSecondRuleModule = ByMillisecondOfSecondRuleModule;
  exports.ByMinuteOfHourRule = ByMinuteOfHourRule;
  exports.ByMinuteOfHourRuleModule = ByMinuteOfHourRuleModule;
  exports.ByMonthOfYearRule = ByMonthOfYearRule;
  exports.ByMonthOfYearRuleModule = ByMonthOfYearRuleModule;
  exports.BySecondOfMinuteRule = BySecondOfMinuteRule;
  exports.BySecondOfMinuteRuleModule = BySecondOfMinuteRuleModule;
  exports.FrequencyRule = FrequencyRule;
  exports.FrequencyRuleModule = FrequencyRuleModule;
  exports.ICAL_RULES = ICAL_RULES;
  exports.RevByDayOfMonthRule = RevByDayOfMonthRule;
  exports.RevByDayOfWeekRule = RevByDayOfWeekRule;
  exports.RevByHourOfDayRule = RevByHourOfDayRule;
  exports.RevByMillisecondOfSecondRule = RevByMillisecondOfSecondRule;
  exports.RevByMinuteOfHourRule = RevByMinuteOfHourRule;
  exports.RevByMonthOfYearRule = RevByMonthOfYearRule;
  exports.RevBySecondOfMinuteRule = RevBySecondOfMinuteRule;
  exports.RevFrequencyRule = RevFrequencyRule;
  exports.getPrevWeekdaysOfMonth = getPrevWeekdaysOfMonth;
  exports.getPrevWeekdaysOfYear = getPrevWeekdaysOfYear;

  Object.defineProperty(exports, '__esModule', { value: true });
});
