// @rschedule/standard-date-adapter/umd/main.js
(function(global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined'
    ? factory(exports, require('@rschedule/core'))
    : typeof define === 'function' && define.amd
    ? define(['exports', '@rschedule/core'], factory)
    : ((global = global || self),
      factory((global.rScheduleStandardDateAdapter = {}), global.rSchedule));
})(this, function(exports, core) {
  'use strict';

  /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
  /* global Reflect, Promise */

  var extendStatics = function(d, b) {
    extendStatics =
      Object.setPrototypeOf ||
      ({ __proto__: [] } instanceof Array &&
        function(d, b) {
          d.__proto__ = b;
        }) ||
      function(d, b) {
        for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
      };
    return extendStatics(d, b);
  };

  function __extends(d, b) {
    extendStatics(d, b);
    function __() {
      this.constructor = d;
    }
    d.prototype = b === null ? Object.create(b) : ((__.prototype = b.prototype), new __());
  }

  function __read(o, n) {
    var m = typeof Symbol === 'function' && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o),
      r,
      ar = [],
      e;
    try {
      while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    } catch (error) {
      e = { error: error };
    } finally {
      try {
        if (r && !r.done && (m = i['return'])) m.call(i);
      } finally {
        if (e) throw e.error;
      }
    }
    return ar;
  }

  function __spread() {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
  }

  var StandardDateAdapter = /** @class */ (function(_super) {
    __extends(StandardDateAdapter, _super);
    function StandardDateAdapter(date, options) {
      if (options === void 0) {
        options = {};
      }
      var _this = _super.call(this, undefined, options) || this;
      if (!['UTC', null, undefined].includes(options.timezone)) {
        throw new core.InvalidDateAdapterError(
          'StandardDateAdapter only supports "UTC" and ' +
            ('local time zones but "' + options.timezone + '" was provided.'),
        );
      }
      _this._date = new Date(date);
      _this.timezone = options.timezone !== undefined ? options.timezone : null;
      if (options.metadata) {
        Object.assign(_this.metadata, options.metadata);
      }
      _this.assertIsValid();
      return _this;
    }
    StandardDateAdapter.isDate = function(object) {
      return Object.prototype.toString.call(object) === '[object Date]';
    };
    StandardDateAdapter.fromDate = function(date, options) {
      return new StandardDateAdapter(date, options);
    };
    StandardDateAdapter.fromJSON = function(json) {
      var args = [
        json.year,
        json.month - 1,
        json.day,
        json.hour,
        json.minute,
        json.second,
        json.millisecond,
      ];
      var date;
      switch (json.timezone) {
        case 'UTC': {
          date = new StandardDateAdapter(new Date(Date.UTC.apply(Date, __spread(args))), {
            timezone: 'UTC',
            duration: json.duration,
          });
          break;
        }
        case null: {
          date = new StandardDateAdapter(new (Date.bind.apply(Date, __spread([void 0], args)))(), {
            timezone: null,
            duration: json.duration,
          });
          break;
        }
        default:
          throw new core.InvalidDateAdapterError(
            'The `StandardDateAdapter` only supports datetimes in ' +
              'UTC or LOCAL (null) time. You provided a JSON object ' +
              ('with timezone "' + json.timezone + '".'),
          );
      }
      if (json.metadata) {
        Object.assign(date.metadata, json.metadata);
      }
      return date;
    };
    StandardDateAdapter.fromDateTime = function(datetime) {
      var _a;
      var date = StandardDateAdapter.fromJSON(datetime.toJSON());
      (_a = date.generators).push.apply(_a, __spread(datetime.generators));
      if (datetime.metadata) {
        Object.assign(date.metadata, datetime.metadata);
      }
      return date;
    };
    Object.defineProperty(StandardDateAdapter.prototype, 'date', {
      get: function() {
        return new Date(this._date);
      },
      enumerable: false,
      configurable: true,
    });
    Object.defineProperty(StandardDateAdapter.prototype, 'end', {
      get: function() {
        if (!this.duration) return;
        if (this._end) return this._end;
        this._end = StandardDateAdapter.fromDateTime(
          this.toDateTime().add(this.duration, 'millisecond'),
        ).date;
        return this._end;
      },
      enumerable: false,
      configurable: true,
    });
    StandardDateAdapter.prototype.set = function(prop, value) {
      if (prop === 'timezone') {
        if (this.timezone === value) return this;
        else {
          return new StandardDateAdapter(this._date, {
            timezone: value,
            duration: this.duration,
            generators: this.generators,
          });
        }
      } else if (prop === 'duration') {
        if (this.duration === value) return this;
        else {
          return new StandardDateAdapter(this._date, {
            timezone: this.timezone,
            duration: value,
            generators: this.generators,
          });
        }
      }
      throw new core.ArgumentError('Unknown prop "' + prop + '" for StandardDateAdapter#set()');
    };
    StandardDateAdapter.prototype.valueOf = function() {
      return this._date.valueOf();
    };
    StandardDateAdapter.prototype.toJSON = function() {
      var json;
      if (this.timezone === 'UTC') {
        json = {
          timezone: this.timezone,
          year: this._date.getUTCFullYear(),
          month: this._date.getUTCMonth() + 1,
          day: this._date.getUTCDate(),
          hour: this._date.getUTCHours(),
          minute: this._date.getUTCMinutes(),
          second: this._date.getUTCSeconds(),
          millisecond: this._date.getUTCMilliseconds(),
        };
      } else {
        json = {
          timezone: this.timezone,
          year: this._date.getFullYear(),
          month: this._date.getMonth() + 1,
          day: this._date.getDate(),
          hour: this._date.getHours(),
          minute: this._date.getMinutes(),
          second: this._date.getSeconds(),
          millisecond: this._date.getMilliseconds(),
        };
      }
      if (this.duration) {
        json.duration = this.duration;
      }
      return json;
    };
    StandardDateAdapter.prototype.assertIsValid = function() {
      if (!StandardDateAdapter.isDate(this._date) || isNaN(this._date.valueOf())) {
        throw new core.InvalidDateAdapterError('StandardDateAdapter has invalid date.');
      } else if (![null, 'UTC'].includes(this.timezone)) {
        throw new core.InvalidDateAdapterError(
          'StandardDateAdapter only supports local (null) and UTC timezones but ' +
            ('"' + this.timezone + '" was specified.'),
        );
      } else if (this.duration && this.duration <= 0) {
        throw new core.InvalidDateAdapterError('If provided, duration must be greater than 0.');
      }
      return true;
    };
    StandardDateAdapter.hasTimezoneSupport = false;
    return StandardDateAdapter;
  })(core.DateAdapterBase);

  exports.StandardDateAdapter = StandardDateAdapter;

  Object.defineProperty(exports, '__esModule', { value: true });
});
